{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e661233a-6be9-477a-a2c8-d510dfde4889",
   "metadata": {},
   "source": [
    "# TISCH2에서 데이터 가져오기\n",
    "\n",
    "- Welcome to TISCH2 BETA 190 datasets 6,297,320 cells\n",
    "\n",
    "Tumor Immune Single-cell Hub 2 (TISCH2) is a scRNA-seq database focusing on tumor microenvironment (TME). TISCH2 provides detailed cell-type annotation at the single-cell level, enabling the exploration of TME across different cancer types.\n",
    "\n",
    "접속 주소: http://tisch.comp-genomics.org/gallery/\n",
    "\n",
    "# 필요한 파일\n",
    "\n",
    "- NSCLC_GSE127471_expression.h5\n",
    "- NSCLC_GSE127471_CellMetainfo_table.tsv\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c6390d06-f1f8-409f-998d-13ca51be27d9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "── \u001b[1mAttaching packages\u001b[22m ─────────────────────────────────────── tidyverse 1.3.2 ──\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2\u001b[39m 3.4.1     \u001b[32m✔\u001b[39m \u001b[34mpurrr  \u001b[39m 1.0.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtibble \u001b[39m 3.1.8     \u001b[32m✔\u001b[39m \u001b[34mdplyr  \u001b[39m 1.1.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtidyr  \u001b[39m 1.3.0     \u001b[32m✔\u001b[39m \u001b[34mstringr\u001b[39m 1.5.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mreadr  \u001b[39m 2.1.4     \u001b[32m✔\u001b[39m \u001b[34mforcats\u001b[39m 1.0.0\n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m     masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mflatten_df()\u001b[39m masks \u001b[34mhdf5r\u001b[39m::flatten_df()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m        masks \u001b[34mstats\u001b[39m::lag()\n"
     ]
    }
   ],
   "source": [
    "library(hdf5r)\n",
    "library(tidyverse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "910646db-d2ba-4451-8be9-e9adb9af604f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1mRows: \u001b[22m\u001b[34m1108\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m11\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (8): Cell, Celltype (malignancy), Celltype (major-lineage), Celltype (mi...\n",
      "\u001b[32mdbl\u001b[39m (3): UMAP_1, UMAP_2, Cluster\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Cell)`\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 6 × 1119</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Cell</th><th scope=col>UMAP_1</th><th scope=col>UMAP_2</th><th scope=col>Celltype (malignancy)</th><th scope=col>Celltype (major-lineage)</th><th scope=col>Celltype (minor-lineage)</th><th scope=col>Cluster</th><th scope=col>Patient</th><th scope=col>Sample</th><th scope=col>Source</th><th scope=col>⋯</th><th scope=col>TAAGCGTTCGCCCTTA-1</th><th scope=col>TACTCGCAGGATCGCA-1</th><th scope=col>TCAGCTCGTCTGGAGA-1</th><th scope=col>TCTATTGAGCCCAATT-1</th><th scope=col>TGACGGCTCTAACCGA-1</th><th scope=col>TGAGGGAAGCCGATTT-1</th><th scope=col>TGCCCTAAGCTACCGC-1</th><th scope=col>TGGTTCCGTGTCCTCT-1</th><th scope=col>TTCGGTCGTCCGTGAC-1</th><th scope=col>TTGGAACCATAGGATA-1</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>AACGTTGGTCATACTG-1</td><td>-1.411341</td><td>6.299751</td><td>Immune cells</td><td>B</td><td>B</td><td>7</td><td>P1</td><td>P1</td><td>PBMC</td><td>⋯</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><td>AAGACCTCACAGGCCT-1</td><td>-1.240770</td><td>6.127520</td><td>Immune cells</td><td>B</td><td>B</td><td>7</td><td>P1</td><td>P1</td><td>PBMC</td><td>⋯</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><td>ACCGTAAGTCTTGTCC-1</td><td>-1.207882</td><td>6.265944</td><td>Immune cells</td><td>B</td><td>B</td><td>7</td><td>P1</td><td>P1</td><td>PBMC</td><td>⋯</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><td>ACGCAGCGTGTGCCTG-1</td><td>-1.991852</td><td>6.321991</td><td>Immune cells</td><td>B</td><td>B</td><td>7</td><td>P1</td><td>P1</td><td>PBMC</td><td>⋯</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><td>ACGGAGACAAGCCTAT-1</td><td>-2.926991</td><td>6.180597</td><td>Immune cells</td><td>B</td><td>B</td><td>7</td><td>P1</td><td>P1</td><td>PBMC</td><td>⋯</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><td>ACGGCCATCACCGGGT-1</td><td>-2.652965</td><td>6.081234</td><td>Immune cells</td><td>B</td><td>B</td><td>7</td><td>P1</td><td>P1</td><td>PBMC</td><td>⋯</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 6 × 1119\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " Cell & UMAP\\_1 & UMAP\\_2 & Celltype (malignancy) & Celltype (major-lineage) & Celltype (minor-lineage) & Cluster & Patient & Sample & Source & ⋯ & TAAGCGTTCGCCCTTA-1 & TACTCGCAGGATCGCA-1 & TCAGCTCGTCTGGAGA-1 & TCTATTGAGCCCAATT-1 & TGACGGCTCTAACCGA-1 & TGAGGGAAGCCGATTT-1 & TGCCCTAAGCTACCGC-1 & TGGTTCCGTGTCCTCT-1 & TTCGGTCGTCCGTGAC-1 & TTGGAACCATAGGATA-1\\\\\n",
       " <chr> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <dbl> & <chr> & <chr> & <chr> & ⋯ & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t AACGTTGGTCATACTG-1 & -1.411341 & 6.299751 & Immune cells & B & B & 7 & P1 & P1 & PBMC & ⋯ & NA & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\t AAGACCTCACAGGCCT-1 & -1.240770 & 6.127520 & Immune cells & B & B & 7 & P1 & P1 & PBMC & ⋯ & NA & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\t ACCGTAAGTCTTGTCC-1 & -1.207882 & 6.265944 & Immune cells & B & B & 7 & P1 & P1 & PBMC & ⋯ & NA & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\t ACGCAGCGTGTGCCTG-1 & -1.991852 & 6.321991 & Immune cells & B & B & 7 & P1 & P1 & PBMC & ⋯ & NA & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\t ACGGAGACAAGCCTAT-1 & -2.926991 & 6.180597 & Immune cells & B & B & 7 & P1 & P1 & PBMC & ⋯ & NA & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\t ACGGCCATCACCGGGT-1 & -2.652965 & 6.081234 & Immune cells & B & B & 7 & P1 & P1 & PBMC & ⋯ & NA & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 6 × 1119\n",
       "\n",
       "| Cell &lt;chr&gt; | UMAP_1 &lt;dbl&gt; | UMAP_2 &lt;dbl&gt; | Celltype (malignancy) &lt;chr&gt; | Celltype (major-lineage) &lt;chr&gt; | Celltype (minor-lineage) &lt;chr&gt; | Cluster &lt;dbl&gt; | Patient &lt;chr&gt; | Sample &lt;chr&gt; | Source &lt;chr&gt; | ⋯ ⋯ | TAAGCGTTCGCCCTTA-1 &lt;dbl&gt; | TACTCGCAGGATCGCA-1 &lt;dbl&gt; | TCAGCTCGTCTGGAGA-1 &lt;dbl&gt; | TCTATTGAGCCCAATT-1 &lt;dbl&gt; | TGACGGCTCTAACCGA-1 &lt;dbl&gt; | TGAGGGAAGCCGATTT-1 &lt;dbl&gt; | TGCCCTAAGCTACCGC-1 &lt;dbl&gt; | TGGTTCCGTGTCCTCT-1 &lt;dbl&gt; | TTCGGTCGTCCGTGAC-1 &lt;dbl&gt; | TTGGAACCATAGGATA-1 &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| AACGTTGGTCATACTG-1 | -1.411341 | 6.299751 | Immune cells | B | B | 7 | P1 | P1 | PBMC | ⋯ | NA | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| AAGACCTCACAGGCCT-1 | -1.240770 | 6.127520 | Immune cells | B | B | 7 | P1 | P1 | PBMC | ⋯ | NA | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| ACCGTAAGTCTTGTCC-1 | -1.207882 | 6.265944 | Immune cells | B | B | 7 | P1 | P1 | PBMC | ⋯ | NA | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| ACGCAGCGTGTGCCTG-1 | -1.991852 | 6.321991 | Immune cells | B | B | 7 | P1 | P1 | PBMC | ⋯ | NA | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| ACGGAGACAAGCCTAT-1 | -2.926991 | 6.180597 | Immune cells | B | B | 7 | P1 | P1 | PBMC | ⋯ | NA | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| ACGGCCATCACCGGGT-1 | -2.652965 | 6.081234 | Immune cells | B | B | 7 | P1 | P1 | PBMC | ⋯ | NA | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "\n"
      ],
      "text/plain": [
       "  Cell               UMAP_1    UMAP_2   Celltype (malignancy)\n",
       "1 AACGTTGGTCATACTG-1 -1.411341 6.299751 Immune cells         \n",
       "2 AAGACCTCACAGGCCT-1 -1.240770 6.127520 Immune cells         \n",
       "3 ACCGTAAGTCTTGTCC-1 -1.207882 6.265944 Immune cells         \n",
       "4 ACGCAGCGTGTGCCTG-1 -1.991852 6.321991 Immune cells         \n",
       "5 ACGGAGACAAGCCTAT-1 -2.926991 6.180597 Immune cells         \n",
       "6 ACGGCCATCACCGGGT-1 -2.652965 6.081234 Immune cells         \n",
       "  Celltype (major-lineage) Celltype (minor-lineage) Cluster Patient Sample\n",
       "1 B                        B                        7       P1      P1    \n",
       "2 B                        B                        7       P1      P1    \n",
       "3 B                        B                        7       P1      P1    \n",
       "4 B                        B                        7       P1      P1    \n",
       "5 B                        B                        7       P1      P1    \n",
       "6 B                        B                        7       P1      P1    \n",
       "  Source ⋯ TAAGCGTTCGCCCTTA-1 TACTCGCAGGATCGCA-1 TCAGCTCGTCTGGAGA-1\n",
       "1 PBMC   ⋯ NA                 NA                 NA                \n",
       "2 PBMC   ⋯ NA                 NA                 NA                \n",
       "3 PBMC   ⋯ NA                 NA                 NA                \n",
       "4 PBMC   ⋯ NA                 NA                 NA                \n",
       "5 PBMC   ⋯ NA                 NA                 NA                \n",
       "6 PBMC   ⋯ NA                 NA                 NA                \n",
       "  TCTATTGAGCCCAATT-1 TGACGGCTCTAACCGA-1 TGAGGGAAGCCGATTT-1 TGCCCTAAGCTACCGC-1\n",
       "1 NA                 NA                 NA                 NA                \n",
       "2 NA                 NA                 NA                 NA                \n",
       "3 NA                 NA                 NA                 NA                \n",
       "4 NA                 NA                 NA                 NA                \n",
       "5 NA                 NA                 NA                 NA                \n",
       "6 NA                 NA                 NA                 NA                \n",
       "  TGGTTCCGTGTCCTCT-1 TTCGGTCGTCCGTGAC-1 TTGGAACCATAGGATA-1\n",
       "1 NA                 NA                 NA                \n",
       "2 NA                 NA                 NA                \n",
       "3 NA                 NA                 NA                \n",
       "4 NA                 NA                 NA                \n",
       "5 NA                 NA                 NA                \n",
       "6 NA                 NA                 NA                "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "h5_data <- hdf5r::H5File$new('../input/NSCLC_GSE127471_expression.h5', mode = 'r')\n",
    "df_meta <- read_tsv(\"../input/NSCLC_GSE127471_CellMetainfo_table.tsv\")\n",
    "\n",
    "feature_matrix <- Matrix::sparseMatrix(\n",
    "  i = h5_data[['matrix/indices']][],\n",
    "  p = h5_data[['matrix/indptr']][],\n",
    "  x = h5_data[['matrix/data']][],\n",
    "  dimnames = list(\n",
    "    h5_data[['matrix/features/name']][],\n",
    "    h5_data[['matrix/barcodes']][]\n",
    "  ),\n",
    "  dims = h5_data[['matrix/shape']][],\n",
    "  index1 = FALSE\n",
    ")\n",
    "\n",
    "# Specify rows to keep\n",
    "keep_rows <- c(\"TIGIT\",\"PVR\",\"CD226\",\"PVRL2\",\"LILRB1\", \"LILRB2\", \"LILRB4\", \"HLA-G\")         \n",
    "m_subset <- feature_matrix[rownames(feature_matrix) %in% keep_rows, ] # Extract rows from matrix\n",
    "# mat_subset\n",
    "\n",
    "m_subset %>% as.matrix() %>%\n",
    "  as_tibble(t(.), rownames=\"Cell\") -> df_mat\n",
    "\n",
    "# df_mat <- as.data.frame(as.matrix(m_subset))\n",
    "# df_mat <- as_tibble(t(df_mat), rownames=\"Cell\")\n",
    "df <- df_meta %>% left_join(., df_mat)\n",
    "head(df)\n",
    "\n",
    "#write.csv(t(df), \"../output/230220_GSE127471.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b6a1b2a-cb58-4473-8e74-5f6301ca4f9f",
   "metadata": {},
   "source": [
    "# Todo\n",
    "\n",
    "함수로 만들어서 반복문을 하면 좋을 것 같다"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b19e9843-2286-490a-848a-76f16587c598",
   "metadata": {},
   "outputs": [],
   "source": [
    "#write.csv(t(df), \"../output/230220_GSE127471.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "af748ce8-5420-4226-b2e9-28cfdb9b012c",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(fs)\n",
    "files <- dir_ls('../input/', glob='*.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "581aeca1-c4dd-4486-98c3-b3ea1020abec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>../input/230119_df_Bernard_Thienpont.csv:</strong> '../input/230119_df_Bernard_Thienpont.csv'"
      ],
      "text/latex": [
       "\\textbf{../input/230119\\textbackslash{}\\_df\\textbackslash{}\\_Bernard\\textbackslash{}\\_Thienpont.csv:} '../input/230119\\_df\\_Bernard\\_Thienpont.csv'"
      ],
      "text/markdown": [
       "**../input/230119_df_Bernard_Thienpont.csv:** '../input/230119_df_Bernard_Thienpont.csv'"
      ],
      "text/plain": [
       "../input/230119_df_Bernard_Thienpont.csv"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "files[1]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "686411c1-5667-4b80-96ee-d7438584ced4",
   "metadata": {},
   "source": [
    "# cell 마다 데이터베이스 다운로드 하게 만들기\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5f079ae4-1d65-43a5-abe6-9db16fbd34d7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1mRows: \u001b[22m\u001b[34m1108\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m11\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (8): Cell, Celltype (malignancy), Celltype (major-lineage), Celltype (mi...\n",
      "\u001b[32mdbl\u001b[39m (3): UMAP_1, UMAP_2, Cluster\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Cell)`\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 6 × 18</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Cell</th><th scope=col>UMAP_1</th><th scope=col>UMAP_2</th><th scope=col>Celltype (malignancy)</th><th scope=col>Celltype (major-lineage)</th><th scope=col>Celltype (minor-lineage)</th><th scope=col>Cluster</th><th scope=col>Patient</th><th scope=col>Sample</th><th scope=col>Source</th><th scope=col>Stage</th><th scope=col>TIGIT</th><th scope=col>HLA-G</th><th scope=col>CD226</th><th scope=col>PVRL2</th><th scope=col>LILRB2</th><th scope=col>LILRB1</th><th scope=col>LILRB4</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>AACGTTGGTCATACTG-1</td><td>-1.411341</td><td>6.299751</td><td>Immune cells</td><td>B</td><td>B</td><td>7</td><td>P1</td><td>P1</td><td>PBMC</td><td>Primary</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0.000000</td><td>0</td></tr>\n",
       "\t<tr><td>AAGACCTCACAGGCCT-1</td><td>-1.240770</td><td>6.127520</td><td>Immune cells</td><td>B</td><td>B</td><td>7</td><td>P1</td><td>P1</td><td>PBMC</td><td>Primary</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0.000000</td><td>0</td></tr>\n",
       "\t<tr><td>ACCGTAAGTCTTGTCC-1</td><td>-1.207882</td><td>6.265944</td><td>Immune cells</td><td>B</td><td>B</td><td>7</td><td>P1</td><td>P1</td><td>PBMC</td><td>Primary</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0.000000</td><td>0</td></tr>\n",
       "\t<tr><td>ACGCAGCGTGTGCCTG-1</td><td>-1.991852</td><td>6.321991</td><td>Immune cells</td><td>B</td><td>B</td><td>7</td><td>P1</td><td>P1</td><td>PBMC</td><td>Primary</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0.000000</td><td>0</td></tr>\n",
       "\t<tr><td>ACGGAGACAAGCCTAT-1</td><td>-2.926991</td><td>6.180597</td><td>Immune cells</td><td>B</td><td>B</td><td>7</td><td>P1</td><td>P1</td><td>PBMC</td><td>Primary</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>1.062495</td><td>0</td></tr>\n",
       "\t<tr><td>ACGGCCATCACCGGGT-1</td><td>-2.652965</td><td>6.081234</td><td>Immune cells</td><td>B</td><td>B</td><td>7</td><td>P1</td><td>P1</td><td>PBMC</td><td>Primary</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0.000000</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 6 × 18\n",
       "\\begin{tabular}{llllllllllllllllll}\n",
       " Cell & UMAP\\_1 & UMAP\\_2 & Celltype (malignancy) & Celltype (major-lineage) & Celltype (minor-lineage) & Cluster & Patient & Sample & Source & Stage & TIGIT & HLA-G & CD226 & PVRL2 & LILRB2 & LILRB1 & LILRB4\\\\\n",
       " <chr> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <dbl> & <chr> & <chr> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t AACGTTGGTCATACTG-1 & -1.411341 & 6.299751 & Immune cells & B & B & 7 & P1 & P1 & PBMC & Primary & 0 & 0 & 0 & 0 & 0 & 0.000000 & 0\\\\\n",
       "\t AAGACCTCACAGGCCT-1 & -1.240770 & 6.127520 & Immune cells & B & B & 7 & P1 & P1 & PBMC & Primary & 0 & 0 & 0 & 0 & 0 & 0.000000 & 0\\\\\n",
       "\t ACCGTAAGTCTTGTCC-1 & -1.207882 & 6.265944 & Immune cells & B & B & 7 & P1 & P1 & PBMC & Primary & 0 & 0 & 0 & 0 & 0 & 0.000000 & 0\\\\\n",
       "\t ACGCAGCGTGTGCCTG-1 & -1.991852 & 6.321991 & Immune cells & B & B & 7 & P1 & P1 & PBMC & Primary & 0 & 0 & 0 & 0 & 0 & 0.000000 & 0\\\\\n",
       "\t ACGGAGACAAGCCTAT-1 & -2.926991 & 6.180597 & Immune cells & B & B & 7 & P1 & P1 & PBMC & Primary & 0 & 0 & 0 & 0 & 0 & 1.062495 & 0\\\\\n",
       "\t ACGGCCATCACCGGGT-1 & -2.652965 & 6.081234 & Immune cells & B & B & 7 & P1 & P1 & PBMC & Primary & 0 & 0 & 0 & 0 & 0 & 0.000000 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 6 × 18\n",
       "\n",
       "| Cell &lt;chr&gt; | UMAP_1 &lt;dbl&gt; | UMAP_2 &lt;dbl&gt; | Celltype (malignancy) &lt;chr&gt; | Celltype (major-lineage) &lt;chr&gt; | Celltype (minor-lineage) &lt;chr&gt; | Cluster &lt;dbl&gt; | Patient &lt;chr&gt; | Sample &lt;chr&gt; | Source &lt;chr&gt; | Stage &lt;chr&gt; | TIGIT &lt;dbl&gt; | HLA-G &lt;dbl&gt; | CD226 &lt;dbl&gt; | PVRL2 &lt;dbl&gt; | LILRB2 &lt;dbl&gt; | LILRB1 &lt;dbl&gt; | LILRB4 &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| AACGTTGGTCATACTG-1 | -1.411341 | 6.299751 | Immune cells | B | B | 7 | P1 | P1 | PBMC | Primary | 0 | 0 | 0 | 0 | 0 | 0.000000 | 0 |\n",
       "| AAGACCTCACAGGCCT-1 | -1.240770 | 6.127520 | Immune cells | B | B | 7 | P1 | P1 | PBMC | Primary | 0 | 0 | 0 | 0 | 0 | 0.000000 | 0 |\n",
       "| ACCGTAAGTCTTGTCC-1 | -1.207882 | 6.265944 | Immune cells | B | B | 7 | P1 | P1 | PBMC | Primary | 0 | 0 | 0 | 0 | 0 | 0.000000 | 0 |\n",
       "| ACGCAGCGTGTGCCTG-1 | -1.991852 | 6.321991 | Immune cells | B | B | 7 | P1 | P1 | PBMC | Primary | 0 | 0 | 0 | 0 | 0 | 0.000000 | 0 |\n",
       "| ACGGAGACAAGCCTAT-1 | -2.926991 | 6.180597 | Immune cells | B | B | 7 | P1 | P1 | PBMC | Primary | 0 | 0 | 0 | 0 | 0 | 1.062495 | 0 |\n",
       "| ACGGCCATCACCGGGT-1 | -2.652965 | 6.081234 | Immune cells | B | B | 7 | P1 | P1 | PBMC | Primary | 0 | 0 | 0 | 0 | 0 | 0.000000 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "  Cell               UMAP_1    UMAP_2   Celltype (malignancy)\n",
       "1 AACGTTGGTCATACTG-1 -1.411341 6.299751 Immune cells         \n",
       "2 AAGACCTCACAGGCCT-1 -1.240770 6.127520 Immune cells         \n",
       "3 ACCGTAAGTCTTGTCC-1 -1.207882 6.265944 Immune cells         \n",
       "4 ACGCAGCGTGTGCCTG-1 -1.991852 6.321991 Immune cells         \n",
       "5 ACGGAGACAAGCCTAT-1 -2.926991 6.180597 Immune cells         \n",
       "6 ACGGCCATCACCGGGT-1 -2.652965 6.081234 Immune cells         \n",
       "  Celltype (major-lineage) Celltype (minor-lineage) Cluster Patient Sample\n",
       "1 B                        B                        7       P1      P1    \n",
       "2 B                        B                        7       P1      P1    \n",
       "3 B                        B                        7       P1      P1    \n",
       "4 B                        B                        7       P1      P1    \n",
       "5 B                        B                        7       P1      P1    \n",
       "6 B                        B                        7       P1      P1    \n",
       "  Source Stage   TIGIT HLA-G CD226 PVRL2 LILRB2 LILRB1   LILRB4\n",
       "1 PBMC   Primary 0     0     0     0     0      0.000000 0     \n",
       "2 PBMC   Primary 0     0     0     0     0      0.000000 0     \n",
       "3 PBMC   Primary 0     0     0     0     0      0.000000 0     \n",
       "4 PBMC   Primary 0     0     0     0     0      0.000000 0     \n",
       "5 PBMC   Primary 0     0     0     0     0      1.062495 0     \n",
       "6 PBMC   Primary 0     0     0     0     0      0.000000 0     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name <- \"NSCLC_GSE127471\"\n",
    "# Specify rows to keep\n",
    "keep_rows <- c(\"TIGIT\",\"PVR\",\"CD226\",\"PVRL2\",\"LILRB1\", \"LILRB2\", \"LILRB4\", \"HLA-G\")         \n",
    "\n",
    "# load dataset\n",
    "df_meta <- read_tsv(paste(\"../input/\",file_name,\"_CellMetainfo_table.tsv\", sep=\"\"))\n",
    "h5_data <- hdf5r::H5File$new(paste(\"../input/\",file_name,\"_expression.h5\", sep=\"\"), mode = 'r')\n",
    "\n",
    "feature_matrix <- Matrix::sparseMatrix(\n",
    "  i = h5_data[['matrix/indices']][],\n",
    "  p = h5_data[['matrix/indptr']][],\n",
    "  x = h5_data[['matrix/data']][],\n",
    "  dimnames = list(\n",
    "    h5_data[['matrix/features/name']][],\n",
    "    h5_data[['matrix/barcodes']][]\n",
    "  ),\n",
    "  dims = h5_data[['matrix/shape']][],\n",
    "  index1 = FALSE\n",
    ")\n",
    "rm(h5_data) # for saving memory\n",
    "\n",
    "\n",
    "m_subset <- feature_matrix[rownames(feature_matrix) %in% keep_rows, ] # Extract rows from matrix\n",
    "rm(feature_matrix)\n",
    "\n",
    "df_mat <- as.data.frame(as.matrix(m_subset))\n",
    "df_mat <- as_tibble(t(df_mat), rownames=\"Cell\")\n",
    "\n",
    "df <- df_meta %>% left_join(., df_mat)\n",
    "head(df)\n",
    "\n",
    "# to save files\n",
    "write.csv(df, paste(\"../output/\",file_name,\".csv\", sep=\"\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "301b0790-d87c-48e5-9a34-3ccf98c8db93",
   "metadata": {},
   "source": [
    "위에서 복사한 셀을 아래에 붙여 넣기"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c4984695-ef52-4300-9f09-ec64f981f8ad",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1mRows: \u001b[22m\u001b[34m16291\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m14\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (11): Cell, Celltype (malignancy), Celltype (major-lineage), Celltype (m...\n",
      "\u001b[32mdbl\u001b[39m  (3): UMAP_1, UMAP_2, Cluster\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Cell)`\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 6 × 21</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Cell</th><th scope=col>UMAP_1</th><th scope=col>UMAP_2</th><th scope=col>Celltype (malignancy)</th><th scope=col>Celltype (major-lineage)</th><th scope=col>Celltype (minor-lineage)</th><th scope=col>Cluster</th><th scope=col>Response</th><th scope=col>Treatment</th><th scope=col>Patient</th><th scope=col>⋯</th><th scope=col>Gender</th><th scope=col>Stage</th><th scope=col>TimePoint</th><th scope=col>PVR</th><th scope=col>LILRB1</th><th scope=col>LILRB2</th><th scope=col>CD226</th><th scope=col>TIGIT</th><th scope=col>LILRB4</th><th scope=col>HLA-G</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>⋯</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>B3_P5_M15      </td><td>12.26936</td><td> 2.67050714</td><td>Immune cells</td><td>B</td><td>B</td><td>3</td><td>Responder    </td><td>anti-PD1</td><td>P1</td><td>⋯</td><td>Male</td><td>Metastatic</td><td>Post</td><td>0</td><td>1.3476740</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0.16553912</td></tr>\n",
       "\t<tr><td>B6_P6_M15      </td><td>12.90418</td><td>-0.12057095</td><td>Immune cells</td><td>B</td><td>B</td><td>3</td><td>Responder    </td><td>anti-PD1</td><td>P1</td><td>⋯</td><td>Male</td><td>Metastatic</td><td>Post</td><td>0</td><td>0.5500230</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0.15967263</td></tr>\n",
       "\t<tr><td>D7_P4_M15      </td><td>12.39708</td><td> 2.98989935</td><td>Immune cells</td><td>B</td><td>B</td><td>3</td><td>Responder    </td><td>anti-PD1</td><td>P1</td><td>⋯</td><td>Male</td><td>Metastatic</td><td>Post</td><td>0</td><td>0.9109808</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0.16296770</td></tr>\n",
       "\t<tr><td>A12_P3_M55_L001</td><td>13.27240</td><td>-0.02708223</td><td>Immune cells</td><td>B</td><td>B</td><td>3</td><td>Non-responder</td><td>anti-PD1</td><td>P1</td><td>⋯</td><td>Male</td><td>Metastatic</td><td>Post</td><td>0</td><td>0.0000000</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0.09291377</td></tr>\n",
       "\t<tr><td>A12_P4_M55_L001</td><td>11.99764</td><td> 1.58152444</td><td>Immune cells</td><td>B</td><td>B</td><td>3</td><td>Non-responder</td><td>anti-PD1</td><td>P1</td><td>⋯</td><td>Male</td><td>Metastatic</td><td>Post</td><td>0</td><td>0.0000000</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0.05891580</td></tr>\n",
       "\t<tr><td>A2_P2_M55_L001 </td><td>13.00366</td><td> 0.97222597</td><td>Immune cells</td><td>B</td><td>B</td><td>3</td><td>Non-responder</td><td>anti-PD1</td><td>P1</td><td>⋯</td><td>Male</td><td>Metastatic</td><td>Post</td><td>0</td><td>0.4105434</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0.15003802</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 6 × 21\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " Cell & UMAP\\_1 & UMAP\\_2 & Celltype (malignancy) & Celltype (major-lineage) & Celltype (minor-lineage) & Cluster & Response & Treatment & Patient & ⋯ & Gender & Stage & TimePoint & PVR & LILRB1 & LILRB2 & CD226 & TIGIT & LILRB4 & HLA-G\\\\\n",
       " <chr> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <dbl> & <chr> & <chr> & <chr> & ⋯ & <chr> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t B3\\_P5\\_M15       & 12.26936 &  2.67050714 & Immune cells & B & B & 3 & Responder     & anti-PD1 & P1 & ⋯ & Male & Metastatic & Post & 0 & 1.3476740 & 0 & 0 & 0 & 0 & 0.16553912\\\\\n",
       "\t B6\\_P6\\_M15       & 12.90418 & -0.12057095 & Immune cells & B & B & 3 & Responder     & anti-PD1 & P1 & ⋯ & Male & Metastatic & Post & 0 & 0.5500230 & 0 & 0 & 0 & 0 & 0.15967263\\\\\n",
       "\t D7\\_P4\\_M15       & 12.39708 &  2.98989935 & Immune cells & B & B & 3 & Responder     & anti-PD1 & P1 & ⋯ & Male & Metastatic & Post & 0 & 0.9109808 & 0 & 0 & 0 & 0 & 0.16296770\\\\\n",
       "\t A12\\_P3\\_M55\\_L001 & 13.27240 & -0.02708223 & Immune cells & B & B & 3 & Non-responder & anti-PD1 & P1 & ⋯ & Male & Metastatic & Post & 0 & 0.0000000 & 0 & 0 & 0 & 0 & 0.09291377\\\\\n",
       "\t A12\\_P4\\_M55\\_L001 & 11.99764 &  1.58152444 & Immune cells & B & B & 3 & Non-responder & anti-PD1 & P1 & ⋯ & Male & Metastatic & Post & 0 & 0.0000000 & 0 & 0 & 0 & 0 & 0.05891580\\\\\n",
       "\t A2\\_P2\\_M55\\_L001  & 13.00366 &  0.97222597 & Immune cells & B & B & 3 & Non-responder & anti-PD1 & P1 & ⋯ & Male & Metastatic & Post & 0 & 0.4105434 & 0 & 0 & 0 & 0 & 0.15003802\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 6 × 21\n",
       "\n",
       "| Cell &lt;chr&gt; | UMAP_1 &lt;dbl&gt; | UMAP_2 &lt;dbl&gt; | Celltype (malignancy) &lt;chr&gt; | Celltype (major-lineage) &lt;chr&gt; | Celltype (minor-lineage) &lt;chr&gt; | Cluster &lt;dbl&gt; | Response &lt;chr&gt; | Treatment &lt;chr&gt; | Patient &lt;chr&gt; | ⋯ ⋯ | Gender &lt;chr&gt; | Stage &lt;chr&gt; | TimePoint &lt;chr&gt; | PVR &lt;dbl&gt; | LILRB1 &lt;dbl&gt; | LILRB2 &lt;dbl&gt; | CD226 &lt;dbl&gt; | TIGIT &lt;dbl&gt; | LILRB4 &lt;dbl&gt; | HLA-G &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| B3_P5_M15       | 12.26936 |  2.67050714 | Immune cells | B | B | 3 | Responder     | anti-PD1 | P1 | ⋯ | Male | Metastatic | Post | 0 | 1.3476740 | 0 | 0 | 0 | 0 | 0.16553912 |\n",
       "| B6_P6_M15       | 12.90418 | -0.12057095 | Immune cells | B | B | 3 | Responder     | anti-PD1 | P1 | ⋯ | Male | Metastatic | Post | 0 | 0.5500230 | 0 | 0 | 0 | 0 | 0.15967263 |\n",
       "| D7_P4_M15       | 12.39708 |  2.98989935 | Immune cells | B | B | 3 | Responder     | anti-PD1 | P1 | ⋯ | Male | Metastatic | Post | 0 | 0.9109808 | 0 | 0 | 0 | 0 | 0.16296770 |\n",
       "| A12_P3_M55_L001 | 13.27240 | -0.02708223 | Immune cells | B | B | 3 | Non-responder | anti-PD1 | P1 | ⋯ | Male | Metastatic | Post | 0 | 0.0000000 | 0 | 0 | 0 | 0 | 0.09291377 |\n",
       "| A12_P4_M55_L001 | 11.99764 |  1.58152444 | Immune cells | B | B | 3 | Non-responder | anti-PD1 | P1 | ⋯ | Male | Metastatic | Post | 0 | 0.0000000 | 0 | 0 | 0 | 0 | 0.05891580 |\n",
       "| A2_P2_M55_L001  | 13.00366 |  0.97222597 | Immune cells | B | B | 3 | Non-responder | anti-PD1 | P1 | ⋯ | Male | Metastatic | Post | 0 | 0.4105434 | 0 | 0 | 0 | 0 | 0.15003802 |\n",
       "\n"
      ],
      "text/plain": [
       "  Cell            UMAP_1   UMAP_2      Celltype (malignancy)\n",
       "1 B3_P5_M15       12.26936  2.67050714 Immune cells         \n",
       "2 B6_P6_M15       12.90418 -0.12057095 Immune cells         \n",
       "3 D7_P4_M15       12.39708  2.98989935 Immune cells         \n",
       "4 A12_P3_M55_L001 13.27240 -0.02708223 Immune cells         \n",
       "5 A12_P4_M55_L001 11.99764  1.58152444 Immune cells         \n",
       "6 A2_P2_M55_L001  13.00366  0.97222597 Immune cells         \n",
       "  Celltype (major-lineage) Celltype (minor-lineage) Cluster Response     \n",
       "1 B                        B                        3       Responder    \n",
       "2 B                        B                        3       Responder    \n",
       "3 B                        B                        3       Responder    \n",
       "4 B                        B                        3       Non-responder\n",
       "5 B                        B                        3       Non-responder\n",
       "6 B                        B                        3       Non-responder\n",
       "  Treatment Patient ⋯ Gender Stage      TimePoint PVR LILRB1    LILRB2 CD226\n",
       "1 anti-PD1  P1      ⋯ Male   Metastatic Post      0   1.3476740 0      0    \n",
       "2 anti-PD1  P1      ⋯ Male   Metastatic Post      0   0.5500230 0      0    \n",
       "3 anti-PD1  P1      ⋯ Male   Metastatic Post      0   0.9109808 0      0    \n",
       "4 anti-PD1  P1      ⋯ Male   Metastatic Post      0   0.0000000 0      0    \n",
       "5 anti-PD1  P1      ⋯ Male   Metastatic Post      0   0.0000000 0      0    \n",
       "6 anti-PD1  P1      ⋯ Male   Metastatic Post      0   0.4105434 0      0    \n",
       "  TIGIT LILRB4 HLA-G     \n",
       "1 0     0      0.16553912\n",
       "2 0     0      0.15967263\n",
       "3 0     0      0.16296770\n",
       "4 0     0      0.09291377\n",
       "5 0     0      0.05891580\n",
       "6 0     0      0.15003802"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name <- \"SKCM_GSE120575_aPD1aCTLA4\"\n",
    "# Specify rows to keep\n",
    "keep_rows <- c(\"TIGIT\",\"PVR\",\"CD226\",\"PVRL2\",\"LILRB1\", \"LILRB2\", \"LILRB4\", \"HLA-G\")         \n",
    "\n",
    "# load dataset\n",
    "df_meta <- read_tsv(paste(\"../input/\",file_name,\"_CellMetainfo_table.tsv\", sep=\"\"))\n",
    "h5_data <- hdf5r::H5File$new(paste(\"../input/\",file_name,\"_expression.h5\", sep=\"\"), mode = 'r')\n",
    "\n",
    "feature_matrix <- Matrix::sparseMatrix(\n",
    "  i = h5_data[['matrix/indices']][],\n",
    "  p = h5_data[['matrix/indptr']][],\n",
    "  x = h5_data[['matrix/data']][],\n",
    "  dimnames = list(\n",
    "    h5_data[['matrix/features/name']][],\n",
    "    h5_data[['matrix/barcodes']][]\n",
    "  ),\n",
    "  dims = h5_data[['matrix/shape']][],\n",
    "  index1 = FALSE\n",
    ")\n",
    "rm(h5_data) # for saving memory\n",
    "\n",
    "\n",
    "m_subset <- feature_matrix[rownames(feature_matrix) %in% keep_rows, ] # Extract rows from matrix\n",
    "rm(feature_matrix)\n",
    "\n",
    "df_mat <- as.data.frame(as.matrix(m_subset))\n",
    "df_mat <- as_tibble(t(df_mat), rownames=\"Cell\")\n",
    "\n",
    "df <- df_meta %>% left_join(., df_mat)\n",
    "head(df)\n",
    "\n",
    "# to save files\n",
    "write.csv(df, paste(\"../output/\",file_name,\".csv\", sep=\"\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "70da0e3e-9d26-45d3-b939-45f1ac138288",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1mRows: \u001b[22m\u001b[34m4645\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m13\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (10): Cell, Celltype (malignancy), Celltype (major-lineage), Celltype (m...\n",
      "\u001b[32mdbl\u001b[39m  (3): UMAP_1, UMAP_2, Cluster\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Cell)`\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 6 × 20</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Cell</th><th scope=col>UMAP_1</th><th scope=col>UMAP_2</th><th scope=col>Celltype (malignancy)</th><th scope=col>Celltype (major-lineage)</th><th scope=col>Celltype (minor-lineage)</th><th scope=col>Celltype (original)</th><th scope=col>Cluster</th><th scope=col>Celltype</th><th scope=col>Patient</th><th scope=col>Source</th><th scope=col>Gender</th><th scope=col>Stage</th><th scope=col>HLA-G</th><th scope=col>LILRB4</th><th scope=col>CD226</th><th scope=col>LILRB1</th><th scope=col>TIGIT</th><th scope=col>LILRB2</th><th scope=col>PVR</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Cy72_CD45_H02_S758_comb</td><td>6.902168</td><td> -9.808507</td><td>Immune cells</td><td>B</td><td>B</td><td>B        </td><td>1</td><td>B        </td><td>Patient72</td><td>Tumor</td><td>Female</td><td>Metastatic</td><td>0.1244284</td><td>0</td><td>0</td><td>0.00000000</td><td>0</td><td>0</td><td>0.00000000</td></tr>\n",
       "\t<tr><td>Cy72_CD45_A07_S679_comb</td><td>7.220452</td><td>-10.508887</td><td>Immune cells</td><td>B</td><td>B</td><td>B        </td><td>1</td><td>B        </td><td>Patient72</td><td>Tumor</td><td>Female</td><td>Metastatic</td><td>0.0434421</td><td>0</td><td>0</td><td>0.00000000</td><td>0</td><td>0</td><td>0.01038673</td></tr>\n",
       "\t<tr><td>Cy71_CD45_H05_S569_comb</td><td>6.830885</td><td>-10.316877</td><td>Immune cells</td><td>B</td><td>B</td><td>Malignant</td><td>1</td><td>Malignant</td><td>Patient71</td><td>Tumor</td><td>Male  </td><td>Metastatic</td><td>0.0000000</td><td>0</td><td>0</td><td>0.14095660</td><td>0</td><td>0</td><td>0.00000000</td></tr>\n",
       "\t<tr><td>Cy74_CD45_B08_S404_comb</td><td>7.106164</td><td>-11.553431</td><td>Immune cells</td><td>B</td><td>B</td><td>B        </td><td>1</td><td>B        </td><td>Patient74</td><td>Tumor</td><td>Male  </td><td>Metastatic</td><td>0.6685714</td><td>0</td><td>0</td><td>0.00000000</td><td>0</td><td>0</td><td>0.00000000</td></tr>\n",
       "\t<tr><td>Cy72_CD45_G09_S753_comb</td><td>6.830161</td><td>-10.858874</td><td>Immune cells</td><td>B</td><td>B</td><td>Malignant</td><td>1</td><td>Malignant</td><td>Patient72</td><td>Tumor</td><td>Female</td><td>Metastatic</td><td>0.0000000</td><td>0</td><td>0</td><td>0.09906139</td><td>0</td><td>0</td><td>0.00000000</td></tr>\n",
       "\t<tr><td>Cy74_CD45_C03_S411_comb</td><td>7.539371</td><td>-10.379721</td><td>Immune cells</td><td>B</td><td>B</td><td>B        </td><td>1</td><td>B        </td><td>Patient74</td><td>Tumor</td><td>Male  </td><td>Metastatic</td><td>0.0000000</td><td>0</td><td>0</td><td>0.00000000</td><td>0</td><td>0</td><td>0.00000000</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 6 × 20\n",
       "\\begin{tabular}{llllllllllllllllllll}\n",
       " Cell & UMAP\\_1 & UMAP\\_2 & Celltype (malignancy) & Celltype (major-lineage) & Celltype (minor-lineage) & Celltype (original) & Cluster & Celltype & Patient & Source & Gender & Stage & HLA-G & LILRB4 & CD226 & LILRB1 & TIGIT & LILRB2 & PVR\\\\\n",
       " <chr> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <chr> & <dbl> & <chr> & <chr> & <chr> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t Cy72\\_CD45\\_H02\\_S758\\_comb & 6.902168 &  -9.808507 & Immune cells & B & B & B         & 1 & B         & Patient72 & Tumor & Female & Metastatic & 0.1244284 & 0 & 0 & 0.00000000 & 0 & 0 & 0.00000000\\\\\n",
       "\t Cy72\\_CD45\\_A07\\_S679\\_comb & 7.220452 & -10.508887 & Immune cells & B & B & B         & 1 & B         & Patient72 & Tumor & Female & Metastatic & 0.0434421 & 0 & 0 & 0.00000000 & 0 & 0 & 0.01038673\\\\\n",
       "\t Cy71\\_CD45\\_H05\\_S569\\_comb & 6.830885 & -10.316877 & Immune cells & B & B & Malignant & 1 & Malignant & Patient71 & Tumor & Male   & Metastatic & 0.0000000 & 0 & 0 & 0.14095660 & 0 & 0 & 0.00000000\\\\\n",
       "\t Cy74\\_CD45\\_B08\\_S404\\_comb & 7.106164 & -11.553431 & Immune cells & B & B & B         & 1 & B         & Patient74 & Tumor & Male   & Metastatic & 0.6685714 & 0 & 0 & 0.00000000 & 0 & 0 & 0.00000000\\\\\n",
       "\t Cy72\\_CD45\\_G09\\_S753\\_comb & 6.830161 & -10.858874 & Immune cells & B & B & Malignant & 1 & Malignant & Patient72 & Tumor & Female & Metastatic & 0.0000000 & 0 & 0 & 0.09906139 & 0 & 0 & 0.00000000\\\\\n",
       "\t Cy74\\_CD45\\_C03\\_S411\\_comb & 7.539371 & -10.379721 & Immune cells & B & B & B         & 1 & B         & Patient74 & Tumor & Male   & Metastatic & 0.0000000 & 0 & 0 & 0.00000000 & 0 & 0 & 0.00000000\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 6 × 20\n",
       "\n",
       "| Cell &lt;chr&gt; | UMAP_1 &lt;dbl&gt; | UMAP_2 &lt;dbl&gt; | Celltype (malignancy) &lt;chr&gt; | Celltype (major-lineage) &lt;chr&gt; | Celltype (minor-lineage) &lt;chr&gt; | Celltype (original) &lt;chr&gt; | Cluster &lt;dbl&gt; | Celltype &lt;chr&gt; | Patient &lt;chr&gt; | Source &lt;chr&gt; | Gender &lt;chr&gt; | Stage &lt;chr&gt; | HLA-G &lt;dbl&gt; | LILRB4 &lt;dbl&gt; | CD226 &lt;dbl&gt; | LILRB1 &lt;dbl&gt; | TIGIT &lt;dbl&gt; | LILRB2 &lt;dbl&gt; | PVR &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| Cy72_CD45_H02_S758_comb | 6.902168 |  -9.808507 | Immune cells | B | B | B         | 1 | B         | Patient72 | Tumor | Female | Metastatic | 0.1244284 | 0 | 0 | 0.00000000 | 0 | 0 | 0.00000000 |\n",
       "| Cy72_CD45_A07_S679_comb | 7.220452 | -10.508887 | Immune cells | B | B | B         | 1 | B         | Patient72 | Tumor | Female | Metastatic | 0.0434421 | 0 | 0 | 0.00000000 | 0 | 0 | 0.01038673 |\n",
       "| Cy71_CD45_H05_S569_comb | 6.830885 | -10.316877 | Immune cells | B | B | Malignant | 1 | Malignant | Patient71 | Tumor | Male   | Metastatic | 0.0000000 | 0 | 0 | 0.14095660 | 0 | 0 | 0.00000000 |\n",
       "| Cy74_CD45_B08_S404_comb | 7.106164 | -11.553431 | Immune cells | B | B | B         | 1 | B         | Patient74 | Tumor | Male   | Metastatic | 0.6685714 | 0 | 0 | 0.00000000 | 0 | 0 | 0.00000000 |\n",
       "| Cy72_CD45_G09_S753_comb | 6.830161 | -10.858874 | Immune cells | B | B | Malignant | 1 | Malignant | Patient72 | Tumor | Female | Metastatic | 0.0000000 | 0 | 0 | 0.09906139 | 0 | 0 | 0.00000000 |\n",
       "| Cy74_CD45_C03_S411_comb | 7.539371 | -10.379721 | Immune cells | B | B | B         | 1 | B         | Patient74 | Tumor | Male   | Metastatic | 0.0000000 | 0 | 0 | 0.00000000 | 0 | 0 | 0.00000000 |\n",
       "\n"
      ],
      "text/plain": [
       "  Cell                    UMAP_1   UMAP_2     Celltype (malignancy)\n",
       "1 Cy72_CD45_H02_S758_comb 6.902168  -9.808507 Immune cells         \n",
       "2 Cy72_CD45_A07_S679_comb 7.220452 -10.508887 Immune cells         \n",
       "3 Cy71_CD45_H05_S569_comb 6.830885 -10.316877 Immune cells         \n",
       "4 Cy74_CD45_B08_S404_comb 7.106164 -11.553431 Immune cells         \n",
       "5 Cy72_CD45_G09_S753_comb 6.830161 -10.858874 Immune cells         \n",
       "6 Cy74_CD45_C03_S411_comb 7.539371 -10.379721 Immune cells         \n",
       "  Celltype (major-lineage) Celltype (minor-lineage) Celltype (original) Cluster\n",
       "1 B                        B                        B                   1      \n",
       "2 B                        B                        B                   1      \n",
       "3 B                        B                        Malignant           1      \n",
       "4 B                        B                        B                   1      \n",
       "5 B                        B                        Malignant           1      \n",
       "6 B                        B                        B                   1      \n",
       "  Celltype  Patient   Source Gender Stage      HLA-G     LILRB4 CD226\n",
       "1 B         Patient72 Tumor  Female Metastatic 0.1244284 0      0    \n",
       "2 B         Patient72 Tumor  Female Metastatic 0.0434421 0      0    \n",
       "3 Malignant Patient71 Tumor  Male   Metastatic 0.0000000 0      0    \n",
       "4 B         Patient74 Tumor  Male   Metastatic 0.6685714 0      0    \n",
       "5 Malignant Patient72 Tumor  Female Metastatic 0.0000000 0      0    \n",
       "6 B         Patient74 Tumor  Male   Metastatic 0.0000000 0      0    \n",
       "  LILRB1     TIGIT LILRB2 PVR       \n",
       "1 0.00000000 0     0      0.00000000\n",
       "2 0.00000000 0     0      0.01038673\n",
       "3 0.14095660 0     0      0.00000000\n",
       "4 0.00000000 0     0      0.00000000\n",
       "5 0.09906139 0     0      0.00000000\n",
       "6 0.00000000 0     0      0.00000000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name <- \"SKCM_GSE72056\"\n",
    "# Specify rows to keep\n",
    "keep_rows <- c(\"TIGIT\",\"PVR\",\"CD226\",\"PVRL2\",\"LILRB1\", \"LILRB2\", \"LILRB4\", \"HLA-G\")         \n",
    "\n",
    "# load datasetB\n",
    "df_meta <- read_tsv(paste(\"../input/\",file_name,\"_CellMetainfo_table.tsv\", sep=\"\"))\n",
    "h5_data <- hdf5r::H5File$new(paste(\"../input/\",file_name,\"_expression.h5\", sep=\"\"), mode = 'r')\n",
    "\n",
    "feature_matrix <- Matrix::sparseMatrix(\n",
    "  i = h5_data[['matrix/indices']][],\n",
    "  p = h5_data[['matrix/indptr']][],\n",
    "  x = h5_data[['matrix/data']][],\n",
    "  dimnames = list(\n",
    "    h5_data[['matrix/features/name']][],\n",
    "    h5_data[['matrix/barcodes']][]\n",
    "  ),\n",
    "  dims = h5_data[['matrix/shape']][],\n",
    "  index1 = FALSE\n",
    ")\n",
    "rm(h5_data) # for saving memory\n",
    "\n",
    "\n",
    "m_subset <- feature_matrix[rownames(feature_matrix) %in% keep_rows, ] # Extract rows from matrix\n",
    "rm(feature_matrix)\n",
    "\n",
    "df_mat <- as.data.frame(as.matrix(m_subset))\n",
    "df_mat <- as_tibble(t(df_mat), rownames=\"Cell\")\n",
    "\n",
    "df <- df_meta %>% left_join(., df_mat)\n",
    "head(df)\n",
    "\n",
    "# to save files\n",
    "write.csv(df, paste(\"../output/\",file_name,\".csv\", sep=\"\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "15a3f07c-a6f1-4dd4-b914-6b8e959a030e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1mRows: \u001b[22m\u001b[34m5902\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m15\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (12): Cell, Celltype (malignancy), Celltype (major-lineage), Celltype (m...\n",
      "\u001b[32mdbl\u001b[39m  (3): UMAP_1, UMAP_2, Cluster\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Cell)`\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 6 × 22</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Cell</th><th scope=col>UMAP_1</th><th scope=col>UMAP_2</th><th scope=col>Celltype (malignancy)</th><th scope=col>Celltype (major-lineage)</th><th scope=col>Celltype (minor-lineage)</th><th scope=col>Celltype (original)</th><th scope=col>Cluster</th><th scope=col>Site</th><th scope=col>Celltype</th><th scope=col>⋯</th><th scope=col>Gender</th><th scope=col>Stage</th><th scope=col>TNMstage</th><th scope=col>HLA-G</th><th scope=col>LILRB4</th><th scope=col>CD226</th><th scope=col>LILRB1</th><th scope=col>TIGIT</th><th scope=col>LILRB2</th><th scope=col>PVR</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>⋯</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>HN25_P25_B08_S308_comb   </td><td> 5.387136</td><td> 4.733791</td><td>Immune cells</td><td>CD4Tconv</td><td>CD4Tn</td><td>Dendritic </td><td>15</td><td>Lymph node</td><td>Dendritic </td><td>⋯</td><td>Female</td><td>Primary</td><td>II </td><td>0.1215125</td><td>0.00000000</td><td>0</td><td>0.0000000</td><td>0.0000000</td><td>0.000000000</td><td>0.000000000</td></tr>\n",
       "\t<tr><td>HN26_P13_A02_S98_comb    </td><td> 5.397628</td><td> 4.681664</td><td>Immune cells</td><td>CD4Tconv</td><td>CD4Tn</td><td>Dendritic </td><td>15</td><td>Lymph node</td><td>Dendritic </td><td>⋯</td><td>Male  </td><td>Primary</td><td>III</td><td>0.4069121</td><td>0.00000000</td><td>0</td><td>0.0000000</td><td>0.0000000</td><td>0.000000000</td><td>0.005795162</td></tr>\n",
       "\t<tr><td>HN26_P26_G10_S370_comb   </td><td> 5.412244</td><td> 4.775695</td><td>Immune cells</td><td>CD4Tconv</td><td>CD4Tn</td><td>Dendritic </td><td>15</td><td>Lymph node</td><td>Dendritic </td><td>⋯</td><td>Male  </td><td>Primary</td><td>III</td><td>0.1650751</td><td>0.00000000</td><td>0</td><td>0.0000000</td><td>0.0000000</td><td>0.000000000</td><td>0.000000000</td></tr>\n",
       "\t<tr><td>HN26_P6_B09_S21_comb     </td><td>10.841489</td><td>-1.543836</td><td>Immune cells</td><td>CD4Tconv</td><td>CD4Tn</td><td>T cell    </td><td> 9</td><td>Primary   </td><td>T cell    </td><td>⋯</td><td>Male  </td><td>Primary</td><td>III</td><td>0.3732727</td><td>0.00000000</td><td>0</td><td>0.0000000</td><td>0.3594962</td><td>0.000000000</td><td>0.000000000</td></tr>\n",
       "\t<tr><td>HNSCC26_P12_E08_S248_comb</td><td> 8.661125</td><td> 5.069691</td><td>Immune cells</td><td>CD4Tconv</td><td>CD4Tn</td><td>Macrophage</td><td>15</td><td>Lymph node</td><td>Macrophage</td><td>⋯</td><td>Male  </td><td>Primary</td><td>III</td><td>0.1154432</td><td>0.01283061</td><td>0</td><td>0.8654621</td><td>0.0000000</td><td>0.005332531</td><td>0.000000000</td></tr>\n",
       "\t<tr><td>HN25_P5_C11_S35_comb     </td><td> 5.380356</td><td> 4.687560</td><td>Immune cells</td><td>CD4Tconv</td><td>CD4Tn</td><td>Dendritic </td><td>15</td><td>Primary   </td><td>Dendritic </td><td>⋯</td><td>Female</td><td>Primary</td><td>II </td><td>0.1856017</td><td>0.00000000</td><td>0</td><td>0.0000000</td><td>0.0000000</td><td>0.000000000</td><td>0.000000000</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 6 × 22\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " Cell & UMAP\\_1 & UMAP\\_2 & Celltype (malignancy) & Celltype (major-lineage) & Celltype (minor-lineage) & Celltype (original) & Cluster & Site & Celltype & ⋯ & Gender & Stage & TNMstage & HLA-G & LILRB4 & CD226 & LILRB1 & TIGIT & LILRB2 & PVR\\\\\n",
       " <chr> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <chr> & <dbl> & <chr> & <chr> & ⋯ & <chr> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t HN25\\_P25\\_B08\\_S308\\_comb    &  5.387136 &  4.733791 & Immune cells & CD4Tconv & CD4Tn & Dendritic  & 15 & Lymph node & Dendritic  & ⋯ & Female & Primary & II  & 0.1215125 & 0.00000000 & 0 & 0.0000000 & 0.0000000 & 0.000000000 & 0.000000000\\\\\n",
       "\t HN26\\_P13\\_A02\\_S98\\_comb     &  5.397628 &  4.681664 & Immune cells & CD4Tconv & CD4Tn & Dendritic  & 15 & Lymph node & Dendritic  & ⋯ & Male   & Primary & III & 0.4069121 & 0.00000000 & 0 & 0.0000000 & 0.0000000 & 0.000000000 & 0.005795162\\\\\n",
       "\t HN26\\_P26\\_G10\\_S370\\_comb    &  5.412244 &  4.775695 & Immune cells & CD4Tconv & CD4Tn & Dendritic  & 15 & Lymph node & Dendritic  & ⋯ & Male   & Primary & III & 0.1650751 & 0.00000000 & 0 & 0.0000000 & 0.0000000 & 0.000000000 & 0.000000000\\\\\n",
       "\t HN26\\_P6\\_B09\\_S21\\_comb      & 10.841489 & -1.543836 & Immune cells & CD4Tconv & CD4Tn & T cell     &  9 & Primary    & T cell     & ⋯ & Male   & Primary & III & 0.3732727 & 0.00000000 & 0 & 0.0000000 & 0.3594962 & 0.000000000 & 0.000000000\\\\\n",
       "\t HNSCC26\\_P12\\_E08\\_S248\\_comb &  8.661125 &  5.069691 & Immune cells & CD4Tconv & CD4Tn & Macrophage & 15 & Lymph node & Macrophage & ⋯ & Male   & Primary & III & 0.1154432 & 0.01283061 & 0 & 0.8654621 & 0.0000000 & 0.005332531 & 0.000000000\\\\\n",
       "\t HN25\\_P5\\_C11\\_S35\\_comb      &  5.380356 &  4.687560 & Immune cells & CD4Tconv & CD4Tn & Dendritic  & 15 & Primary    & Dendritic  & ⋯ & Female & Primary & II  & 0.1856017 & 0.00000000 & 0 & 0.0000000 & 0.0000000 & 0.000000000 & 0.000000000\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 6 × 22\n",
       "\n",
       "| Cell &lt;chr&gt; | UMAP_1 &lt;dbl&gt; | UMAP_2 &lt;dbl&gt; | Celltype (malignancy) &lt;chr&gt; | Celltype (major-lineage) &lt;chr&gt; | Celltype (minor-lineage) &lt;chr&gt; | Celltype (original) &lt;chr&gt; | Cluster &lt;dbl&gt; | Site &lt;chr&gt; | Celltype &lt;chr&gt; | ⋯ ⋯ | Gender &lt;chr&gt; | Stage &lt;chr&gt; | TNMstage &lt;chr&gt; | HLA-G &lt;dbl&gt; | LILRB4 &lt;dbl&gt; | CD226 &lt;dbl&gt; | LILRB1 &lt;dbl&gt; | TIGIT &lt;dbl&gt; | LILRB2 &lt;dbl&gt; | PVR &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| HN25_P25_B08_S308_comb    |  5.387136 |  4.733791 | Immune cells | CD4Tconv | CD4Tn | Dendritic  | 15 | Lymph node | Dendritic  | ⋯ | Female | Primary | II  | 0.1215125 | 0.00000000 | 0 | 0.0000000 | 0.0000000 | 0.000000000 | 0.000000000 |\n",
       "| HN26_P13_A02_S98_comb     |  5.397628 |  4.681664 | Immune cells | CD4Tconv | CD4Tn | Dendritic  | 15 | Lymph node | Dendritic  | ⋯ | Male   | Primary | III | 0.4069121 | 0.00000000 | 0 | 0.0000000 | 0.0000000 | 0.000000000 | 0.005795162 |\n",
       "| HN26_P26_G10_S370_comb    |  5.412244 |  4.775695 | Immune cells | CD4Tconv | CD4Tn | Dendritic  | 15 | Lymph node | Dendritic  | ⋯ | Male   | Primary | III | 0.1650751 | 0.00000000 | 0 | 0.0000000 | 0.0000000 | 0.000000000 | 0.000000000 |\n",
       "| HN26_P6_B09_S21_comb      | 10.841489 | -1.543836 | Immune cells | CD4Tconv | CD4Tn | T cell     |  9 | Primary    | T cell     | ⋯ | Male   | Primary | III | 0.3732727 | 0.00000000 | 0 | 0.0000000 | 0.3594962 | 0.000000000 | 0.000000000 |\n",
       "| HNSCC26_P12_E08_S248_comb |  8.661125 |  5.069691 | Immune cells | CD4Tconv | CD4Tn | Macrophage | 15 | Lymph node | Macrophage | ⋯ | Male   | Primary | III | 0.1154432 | 0.01283061 | 0 | 0.8654621 | 0.0000000 | 0.005332531 | 0.000000000 |\n",
       "| HN25_P5_C11_S35_comb      |  5.380356 |  4.687560 | Immune cells | CD4Tconv | CD4Tn | Dendritic  | 15 | Primary    | Dendritic  | ⋯ | Female | Primary | II  | 0.1856017 | 0.00000000 | 0 | 0.0000000 | 0.0000000 | 0.000000000 | 0.000000000 |\n",
       "\n"
      ],
      "text/plain": [
       "  Cell                      UMAP_1    UMAP_2    Celltype (malignancy)\n",
       "1 HN25_P25_B08_S308_comb     5.387136  4.733791 Immune cells         \n",
       "2 HN26_P13_A02_S98_comb      5.397628  4.681664 Immune cells         \n",
       "3 HN26_P26_G10_S370_comb     5.412244  4.775695 Immune cells         \n",
       "4 HN26_P6_B09_S21_comb      10.841489 -1.543836 Immune cells         \n",
       "5 HNSCC26_P12_E08_S248_comb  8.661125  5.069691 Immune cells         \n",
       "6 HN25_P5_C11_S35_comb       5.380356  4.687560 Immune cells         \n",
       "  Celltype (major-lineage) Celltype (minor-lineage) Celltype (original) Cluster\n",
       "1 CD4Tconv                 CD4Tn                    Dendritic           15     \n",
       "2 CD4Tconv                 CD4Tn                    Dendritic           15     \n",
       "3 CD4Tconv                 CD4Tn                    Dendritic           15     \n",
       "4 CD4Tconv                 CD4Tn                    T cell               9     \n",
       "5 CD4Tconv                 CD4Tn                    Macrophage          15     \n",
       "6 CD4Tconv                 CD4Tn                    Dendritic           15     \n",
       "  Site       Celltype   ⋯ Gender Stage   TNMstage HLA-G     LILRB4     CD226\n",
       "1 Lymph node Dendritic  ⋯ Female Primary II       0.1215125 0.00000000 0    \n",
       "2 Lymph node Dendritic  ⋯ Male   Primary III      0.4069121 0.00000000 0    \n",
       "3 Lymph node Dendritic  ⋯ Male   Primary III      0.1650751 0.00000000 0    \n",
       "4 Primary    T cell     ⋯ Male   Primary III      0.3732727 0.00000000 0    \n",
       "5 Lymph node Macrophage ⋯ Male   Primary III      0.1154432 0.01283061 0    \n",
       "6 Primary    Dendritic  ⋯ Female Primary II       0.1856017 0.00000000 0    \n",
       "  LILRB1    TIGIT     LILRB2      PVR        \n",
       "1 0.0000000 0.0000000 0.000000000 0.000000000\n",
       "2 0.0000000 0.0000000 0.000000000 0.005795162\n",
       "3 0.0000000 0.0000000 0.000000000 0.000000000\n",
       "4 0.0000000 0.3594962 0.000000000 0.000000000\n",
       "5 0.8654621 0.0000000 0.005332531 0.000000000\n",
       "6 0.0000000 0.0000000 0.000000000 0.000000000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name <- \"HNSC_GSE103322\"\n",
    "# Specify rows to keep\n",
    "keep_rows <- c(\"TIGIT\",\"PVR\",\"CD226\",\"PVRL2\",\"LILRB1\", \"LILRB2\", \"LILRB4\", \"HLA-G\")         \n",
    "\n",
    "# load dataset\n",
    "df_meta <- read_tsv(paste(\"../input/\",file_name,\"_CellMetainfo_table.tsv\", sep=\"\"))\n",
    "h5_data <- hdf5r::H5File$new(paste(\"../input/\",file_name,\"_expression.h5\", sep=\"\"), mode = 'r')\n",
    "\n",
    "feature_matrix <- Matrix::sparseMatrix(\n",
    "  i = h5_data[['matrix/indices']][],\n",
    "  p = h5_data[['matrix/indptr']][],\n",
    "  x = h5_data[['matrix/data']][],\n",
    "  dimnames = list(\n",
    "    h5_data[['matrix/features/name']][],\n",
    "    h5_data[['matrix/barcodes']][]\n",
    "  ),\n",
    "  dims = h5_data[['matrix/shape']][],\n",
    "  index1 = FALSE\n",
    ")\n",
    "rm(h5_data) # for saving memory\n",
    "\n",
    "\n",
    "m_subset <- feature_matrix[rownames(feature_matrix) %in% keep_rows, ] # Extract rows from matrix\n",
    "rm(feature_matrix)\n",
    "\n",
    "df_mat <- as.data.frame(as.matrix(m_subset))\n",
    "df_mat <- as_tibble(t(df_mat), rownames=\"Cell\")\n",
    "\n",
    "df <- df_meta %>% left_join(., df_mat)\n",
    "head(df)\n",
    "\n",
    "# to save files\n",
    "write.csv(df, paste(\"../output/\",file_name,\".csv\", sep=\"\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f39619a1-1b9c-48e8-8b5a-e75a947df06c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1mRows: \u001b[22m\u001b[34m19676\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m13\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (10): Cell, Celltype (malignancy), Celltype (major-lineage), Celltype (m...\n",
      "\u001b[32mdbl\u001b[39m  (3): UMAP_1, UMAP_2, Cluster\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Cell)`\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 6 × 21</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Cell</th><th scope=col>UMAP_1</th><th scope=col>UMAP_2</th><th scope=col>Celltype (malignancy)</th><th scope=col>Celltype (major-lineage)</th><th scope=col>Celltype (minor-lineage)</th><th scope=col>Cluster</th><th scope=col>Sample</th><th scope=col>Patient</th><th scope=col>Source</th><th scope=col>⋯</th><th scope=col>Stage</th><th scope=col>TNMstage</th><th scope=col>HLA-G</th><th scope=col>PVR</th><th scope=col>TIGIT</th><th scope=col>CD226</th><th scope=col>LILRB4</th><th scope=col>LILRB2</th><th scope=col>LILRB1</th><th scope=col>PVRL2</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>⋯</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>BC04BLOOD4_218</td><td>-8.800593</td><td>14.55176</td><td>Immune cells</td><td>B</td><td>B</td><td>10</td><td>BC04BLOOD4</td><td>BC04</td><td>PBMC</td><td>⋯</td><td>Primary</td><td>I</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>BC04BLOOD4_272</td><td>-8.305215</td><td>11.86706</td><td>Immune cells</td><td>B</td><td>B</td><td>15</td><td>BC04BLOOD4</td><td>BC04</td><td>PBMC</td><td>⋯</td><td>Primary</td><td>I</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>BC04BLOOD4_290</td><td>-7.171502</td><td>12.08958</td><td>Immune cells</td><td>B</td><td>B</td><td>15</td><td>BC04BLOOD4</td><td>BC04</td><td>PBMC</td><td>⋯</td><td>Primary</td><td>I</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>BC04BLOOD4_340</td><td>-8.384596</td><td>12.34753</td><td>Immune cells</td><td>B</td><td>B</td><td>15</td><td>BC04BLOOD4</td><td>BC04</td><td>PBMC</td><td>⋯</td><td>Primary</td><td>I</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>BC04BLOOD4_498</td><td>-7.047964</td><td>12.00084</td><td>Immune cells</td><td>B</td><td>B</td><td>15</td><td>BC04BLOOD4</td><td>BC04</td><td>PBMC</td><td>⋯</td><td>Primary</td><td>I</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>BC04BLOOD4_715</td><td>-8.249105</td><td>12.27620</td><td>Immune cells</td><td>B</td><td>B</td><td>15</td><td>BC04BLOOD4</td><td>BC04</td><td>PBMC</td><td>⋯</td><td>Primary</td><td>I</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 6 × 21\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " Cell & UMAP\\_1 & UMAP\\_2 & Celltype (malignancy) & Celltype (major-lineage) & Celltype (minor-lineage) & Cluster & Sample & Patient & Source & ⋯ & Stage & TNMstage & HLA-G & PVR & TIGIT & CD226 & LILRB4 & LILRB2 & LILRB1 & PVRL2\\\\\n",
       " <chr> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <dbl> & <chr> & <chr> & <chr> & ⋯ & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t BC04BLOOD4\\_218 & -8.800593 & 14.55176 & Immune cells & B & B & 10 & BC04BLOOD4 & BC04 & PBMC & ⋯ & Primary & I & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\t BC04BLOOD4\\_272 & -8.305215 & 11.86706 & Immune cells & B & B & 15 & BC04BLOOD4 & BC04 & PBMC & ⋯ & Primary & I & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\t BC04BLOOD4\\_290 & -7.171502 & 12.08958 & Immune cells & B & B & 15 & BC04BLOOD4 & BC04 & PBMC & ⋯ & Primary & I & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\t BC04BLOOD4\\_340 & -8.384596 & 12.34753 & Immune cells & B & B & 15 & BC04BLOOD4 & BC04 & PBMC & ⋯ & Primary & I & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\t BC04BLOOD4\\_498 & -7.047964 & 12.00084 & Immune cells & B & B & 15 & BC04BLOOD4 & BC04 & PBMC & ⋯ & Primary & I & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\t BC04BLOOD4\\_715 & -8.249105 & 12.27620 & Immune cells & B & B & 15 & BC04BLOOD4 & BC04 & PBMC & ⋯ & Primary & I & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 6 × 21\n",
       "\n",
       "| Cell &lt;chr&gt; | UMAP_1 &lt;dbl&gt; | UMAP_2 &lt;dbl&gt; | Celltype (malignancy) &lt;chr&gt; | Celltype (major-lineage) &lt;chr&gt; | Celltype (minor-lineage) &lt;chr&gt; | Cluster &lt;dbl&gt; | Sample &lt;chr&gt; | Patient &lt;chr&gt; | Source &lt;chr&gt; | ⋯ ⋯ | Stage &lt;chr&gt; | TNMstage &lt;chr&gt; | HLA-G &lt;dbl&gt; | PVR &lt;dbl&gt; | TIGIT &lt;dbl&gt; | CD226 &lt;dbl&gt; | LILRB4 &lt;dbl&gt; | LILRB2 &lt;dbl&gt; | LILRB1 &lt;dbl&gt; | PVRL2 &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| BC04BLOOD4_218 | -8.800593 | 14.55176 | Immune cells | B | B | 10 | BC04BLOOD4 | BC04 | PBMC | ⋯ | Primary | I | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "| BC04BLOOD4_272 | -8.305215 | 11.86706 | Immune cells | B | B | 15 | BC04BLOOD4 | BC04 | PBMC | ⋯ | Primary | I | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "| BC04BLOOD4_290 | -7.171502 | 12.08958 | Immune cells | B | B | 15 | BC04BLOOD4 | BC04 | PBMC | ⋯ | Primary | I | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "| BC04BLOOD4_340 | -8.384596 | 12.34753 | Immune cells | B | B | 15 | BC04BLOOD4 | BC04 | PBMC | ⋯ | Primary | I | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "| BC04BLOOD4_498 | -7.047964 | 12.00084 | Immune cells | B | B | 15 | BC04BLOOD4 | BC04 | PBMC | ⋯ | Primary | I | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "| BC04BLOOD4_715 | -8.249105 | 12.27620 | Immune cells | B | B | 15 | BC04BLOOD4 | BC04 | PBMC | ⋯ | Primary | I | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "  Cell           UMAP_1    UMAP_2   Celltype (malignancy)\n",
       "1 BC04BLOOD4_218 -8.800593 14.55176 Immune cells         \n",
       "2 BC04BLOOD4_272 -8.305215 11.86706 Immune cells         \n",
       "3 BC04BLOOD4_290 -7.171502 12.08958 Immune cells         \n",
       "4 BC04BLOOD4_340 -8.384596 12.34753 Immune cells         \n",
       "5 BC04BLOOD4_498 -7.047964 12.00084 Immune cells         \n",
       "6 BC04BLOOD4_715 -8.249105 12.27620 Immune cells         \n",
       "  Celltype (major-lineage) Celltype (minor-lineage) Cluster Sample     Patient\n",
       "1 B                        B                        10      BC04BLOOD4 BC04   \n",
       "2 B                        B                        15      BC04BLOOD4 BC04   \n",
       "3 B                        B                        15      BC04BLOOD4 BC04   \n",
       "4 B                        B                        15      BC04BLOOD4 BC04   \n",
       "5 B                        B                        15      BC04BLOOD4 BC04   \n",
       "6 B                        B                        15      BC04BLOOD4 BC04   \n",
       "  Source ⋯ Stage   TNMstage HLA-G PVR TIGIT CD226 LILRB4 LILRB2 LILRB1 PVRL2\n",
       "1 PBMC   ⋯ Primary I        0     0   0     0     0      0      0      0    \n",
       "2 PBMC   ⋯ Primary I        0     0   0     0     0      0      0      0    \n",
       "3 PBMC   ⋯ Primary I        0     0   0     0     0      0      0      0    \n",
       "4 PBMC   ⋯ Primary I        0     0   0     0     0      0      0      0    \n",
       "5 PBMC   ⋯ Primary I        0     0   0     0     0      0      0      0    \n",
       "6 PBMC   ⋯ Primary I        0     0   0     0     0      0      0      0    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name <- \"BRCA_GSE114727_inDrop\"\n",
    "# Specify rows to keep\n",
    "keep_rows <- c(\"TIGIT\",\"PVR\",\"CD226\",\"PVRL2\",\"LILRB1\", \"LILRB2\", \"LILRB4\", \"HLA-G\")         \n",
    "\n",
    "# load dataset\n",
    "df_meta <- read_tsv(paste(\"../input/\",file_name,\"_CellMetainfo_table.tsv\", sep=\"\"))\n",
    "h5_data <- hdf5r::H5File$new(paste(\"../input/\",file_name,\"_expression.h5\", sep=\"\"), mode = 'r')\n",
    "\n",
    "feature_matrix <- Matrix::sparseMatrix(\n",
    "  i = h5_data[['matrix/indices']][],\n",
    "  p = h5_data[['matrix/indptr']][],\n",
    "  x = h5_data[['matrix/data']][],\n",
    "  dimnames = list(\n",
    "    h5_data[['matrix/features/name']][],\n",
    "    h5_data[['matrix/barcodes']][]\n",
    "  ),\n",
    "  dims = h5_data[['matrix/shape']][],\n",
    "  index1 = FALSE\n",
    ")\n",
    "rm(h5_data) # for saving memory\n",
    "\n",
    "\n",
    "m_subset <- feature_matrix[rownames(feature_matrix) %in% keep_rows, ] # Extract rows from matrix\n",
    "rm(feature_matrix)\n",
    "\n",
    "df_mat <- as.data.frame(as.matrix(m_subset))\n",
    "df_mat <- as_tibble(t(df_mat), rownames=\"Cell\")\n",
    "\n",
    "df <- df_meta %>% left_join(., df_mat)\n",
    "head(df)\n",
    "\n",
    "# to save files\n",
    "write.csv(df, paste(\"../output/\",file_name,\".csv\", sep=\"\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d34454f8-ec31-453f-b6b8-50c92166ce9e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1mRows: \u001b[22m\u001b[34m28678\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m13\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (10): Cell, Celltype (malignancy), Celltype (major-lineage), Celltype (m...\n",
      "\u001b[32mdbl\u001b[39m  (3): UMAP_1, UMAP_2, Cluster\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Cell)`\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 6 × 19</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Cell</th><th scope=col>UMAP_1</th><th scope=col>UMAP_2</th><th scope=col>Celltype (malignancy)</th><th scope=col>Celltype (major-lineage)</th><th scope=col>Celltype (minor-lineage)</th><th scope=col>Cluster</th><th scope=col>Patient</th><th scope=col>Sample</th><th scope=col>Source</th><th scope=col>Gender</th><th scope=col>Stage</th><th scope=col>TNMstage</th><th scope=col>TIGIT</th><th scope=col>HLA-G</th><th scope=col>CD226</th><th scope=col>PVR</th><th scope=col>PVRL2</th><th scope=col>LILRB1</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>BC09_TUMOR1@AAACCTGAGTTAAGTG-1</td><td> 2.230674</td><td>2.0426747</td><td>Immune cells</td><td>CD4Tconv</td><td>Th17 </td><td>0</td><td>BC09</td><td>BC09_TUMOR1</td><td>Tumor</td><td>Female</td><td>Primary</td><td>II</td><td>0.000000</td><td>0</td><td>0.0000000</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>BC09_TUMOR1@AAACCTGCAACACCCG-1</td><td> 2.157552</td><td>5.5871748</td><td>Immune cells</td><td>CD4Tconv</td><td>CD4Tn</td><td>1</td><td>BC09</td><td>BC09_TUMOR1</td><td>Tumor</td><td>Female</td><td>Primary</td><td>II</td><td>0.000000</td><td>0</td><td>0.9804128</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>BC09_TUMOR1@AAACCTGTCCCTGACT-1</td><td> 1.537627</td><td>3.9877924</td><td>Immune cells</td><td>CD4Tconv</td><td>CD4Tn</td><td>1</td><td>BC09</td><td>BC09_TUMOR1</td><td>Tumor</td><td>Female</td><td>Primary</td><td>II</td><td>1.314215</td><td>0</td><td>0.0000000</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>BC09_TUMOR1@AAACGGGAGCATGGCA-1</td><td> 4.824210</td><td>2.1798624</td><td>Immune cells</td><td>CD4Tconv</td><td>Th17 </td><td>0</td><td>BC09</td><td>BC09_TUMOR1</td><td>Tumor</td><td>Female</td><td>Primary</td><td>II</td><td>0.000000</td><td>0</td><td>0.0000000</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>BC09_TUMOR1@AAACGGGCACTAAGTC-1</td><td>-7.379394</td><td>0.1972042</td><td>Immune cells</td><td>CD4Tconv</td><td>Tfh  </td><td>6</td><td>BC09</td><td>BC09_TUMOR1</td><td>Tumor</td><td>Female</td><td>Primary</td><td>II</td><td>1.996102</td><td>0</td><td>0.9516879</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>BC09_TUMOR1@AAACGGGCAGCGTTCG-1</td><td> 4.982963</td><td>1.4641977</td><td>Immune cells</td><td>CD4Tconv</td><td>Th17 </td><td>0</td><td>BC09</td><td>BC09_TUMOR1</td><td>Tumor</td><td>Female</td><td>Primary</td><td>II</td><td>0.000000</td><td>0</td><td>0.0000000</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 6 × 19\n",
       "\\begin{tabular}{lllllllllllllllllll}\n",
       " Cell & UMAP\\_1 & UMAP\\_2 & Celltype (malignancy) & Celltype (major-lineage) & Celltype (minor-lineage) & Cluster & Patient & Sample & Source & Gender & Stage & TNMstage & TIGIT & HLA-G & CD226 & PVR & PVRL2 & LILRB1\\\\\n",
       " <chr> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <dbl> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t BC09\\_TUMOR1@AAACCTGAGTTAAGTG-1 &  2.230674 & 2.0426747 & Immune cells & CD4Tconv & Th17  & 0 & BC09 & BC09\\_TUMOR1 & Tumor & Female & Primary & II & 0.000000 & 0 & 0.0000000 & 0 & 0 & 0\\\\\n",
       "\t BC09\\_TUMOR1@AAACCTGCAACACCCG-1 &  2.157552 & 5.5871748 & Immune cells & CD4Tconv & CD4Tn & 1 & BC09 & BC09\\_TUMOR1 & Tumor & Female & Primary & II & 0.000000 & 0 & 0.9804128 & 0 & 0 & 0\\\\\n",
       "\t BC09\\_TUMOR1@AAACCTGTCCCTGACT-1 &  1.537627 & 3.9877924 & Immune cells & CD4Tconv & CD4Tn & 1 & BC09 & BC09\\_TUMOR1 & Tumor & Female & Primary & II & 1.314215 & 0 & 0.0000000 & 0 & 0 & 0\\\\\n",
       "\t BC09\\_TUMOR1@AAACGGGAGCATGGCA-1 &  4.824210 & 2.1798624 & Immune cells & CD4Tconv & Th17  & 0 & BC09 & BC09\\_TUMOR1 & Tumor & Female & Primary & II & 0.000000 & 0 & 0.0000000 & 0 & 0 & 0\\\\\n",
       "\t BC09\\_TUMOR1@AAACGGGCACTAAGTC-1 & -7.379394 & 0.1972042 & Immune cells & CD4Tconv & Tfh   & 6 & BC09 & BC09\\_TUMOR1 & Tumor & Female & Primary & II & 1.996102 & 0 & 0.9516879 & 0 & 0 & 0\\\\\n",
       "\t BC09\\_TUMOR1@AAACGGGCAGCGTTCG-1 &  4.982963 & 1.4641977 & Immune cells & CD4Tconv & Th17  & 0 & BC09 & BC09\\_TUMOR1 & Tumor & Female & Primary & II & 0.000000 & 0 & 0.0000000 & 0 & 0 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 6 × 19\n",
       "\n",
       "| Cell &lt;chr&gt; | UMAP_1 &lt;dbl&gt; | UMAP_2 &lt;dbl&gt; | Celltype (malignancy) &lt;chr&gt; | Celltype (major-lineage) &lt;chr&gt; | Celltype (minor-lineage) &lt;chr&gt; | Cluster &lt;dbl&gt; | Patient &lt;chr&gt; | Sample &lt;chr&gt; | Source &lt;chr&gt; | Gender &lt;chr&gt; | Stage &lt;chr&gt; | TNMstage &lt;chr&gt; | TIGIT &lt;dbl&gt; | HLA-G &lt;dbl&gt; | CD226 &lt;dbl&gt; | PVR &lt;dbl&gt; | PVRL2 &lt;dbl&gt; | LILRB1 &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| BC09_TUMOR1@AAACCTGAGTTAAGTG-1 |  2.230674 | 2.0426747 | Immune cells | CD4Tconv | Th17  | 0 | BC09 | BC09_TUMOR1 | Tumor | Female | Primary | II | 0.000000 | 0 | 0.0000000 | 0 | 0 | 0 |\n",
       "| BC09_TUMOR1@AAACCTGCAACACCCG-1 |  2.157552 | 5.5871748 | Immune cells | CD4Tconv | CD4Tn | 1 | BC09 | BC09_TUMOR1 | Tumor | Female | Primary | II | 0.000000 | 0 | 0.9804128 | 0 | 0 | 0 |\n",
       "| BC09_TUMOR1@AAACCTGTCCCTGACT-1 |  1.537627 | 3.9877924 | Immune cells | CD4Tconv | CD4Tn | 1 | BC09 | BC09_TUMOR1 | Tumor | Female | Primary | II | 1.314215 | 0 | 0.0000000 | 0 | 0 | 0 |\n",
       "| BC09_TUMOR1@AAACGGGAGCATGGCA-1 |  4.824210 | 2.1798624 | Immune cells | CD4Tconv | Th17  | 0 | BC09 | BC09_TUMOR1 | Tumor | Female | Primary | II | 0.000000 | 0 | 0.0000000 | 0 | 0 | 0 |\n",
       "| BC09_TUMOR1@AAACGGGCACTAAGTC-1 | -7.379394 | 0.1972042 | Immune cells | CD4Tconv | Tfh   | 6 | BC09 | BC09_TUMOR1 | Tumor | Female | Primary | II | 1.996102 | 0 | 0.9516879 | 0 | 0 | 0 |\n",
       "| BC09_TUMOR1@AAACGGGCAGCGTTCG-1 |  4.982963 | 1.4641977 | Immune cells | CD4Tconv | Th17  | 0 | BC09 | BC09_TUMOR1 | Tumor | Female | Primary | II | 0.000000 | 0 | 0.0000000 | 0 | 0 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "  Cell                           UMAP_1    UMAP_2    Celltype (malignancy)\n",
       "1 BC09_TUMOR1@AAACCTGAGTTAAGTG-1  2.230674 2.0426747 Immune cells         \n",
       "2 BC09_TUMOR1@AAACCTGCAACACCCG-1  2.157552 5.5871748 Immune cells         \n",
       "3 BC09_TUMOR1@AAACCTGTCCCTGACT-1  1.537627 3.9877924 Immune cells         \n",
       "4 BC09_TUMOR1@AAACGGGAGCATGGCA-1  4.824210 2.1798624 Immune cells         \n",
       "5 BC09_TUMOR1@AAACGGGCACTAAGTC-1 -7.379394 0.1972042 Immune cells         \n",
       "6 BC09_TUMOR1@AAACGGGCAGCGTTCG-1  4.982963 1.4641977 Immune cells         \n",
       "  Celltype (major-lineage) Celltype (minor-lineage) Cluster Patient Sample     \n",
       "1 CD4Tconv                 Th17                     0       BC09    BC09_TUMOR1\n",
       "2 CD4Tconv                 CD4Tn                    1       BC09    BC09_TUMOR1\n",
       "3 CD4Tconv                 CD4Tn                    1       BC09    BC09_TUMOR1\n",
       "4 CD4Tconv                 Th17                     0       BC09    BC09_TUMOR1\n",
       "5 CD4Tconv                 Tfh                      6       BC09    BC09_TUMOR1\n",
       "6 CD4Tconv                 Th17                     0       BC09    BC09_TUMOR1\n",
       "  Source Gender Stage   TNMstage TIGIT    HLA-G CD226     PVR PVRL2 LILRB1\n",
       "1 Tumor  Female Primary II       0.000000 0     0.0000000 0   0     0     \n",
       "2 Tumor  Female Primary II       0.000000 0     0.9804128 0   0     0     \n",
       "3 Tumor  Female Primary II       1.314215 0     0.0000000 0   0     0     \n",
       "4 Tumor  Female Primary II       0.000000 0     0.0000000 0   0     0     \n",
       "5 Tumor  Female Primary II       1.996102 0     0.9516879 0   0     0     \n",
       "6 Tumor  Female Primary II       0.000000 0     0.0000000 0   0     0     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name <- \"BRCA_GSE114727_10X\"\n",
    "# Specify rows to keep\n",
    "keep_rows <- c(\"TIGIT\",\"PVR\",\"CD226\",\"PVRL2\",\"LILRB1\", \"LILRB2\", \"LILRB4\", \"HLA-G\")         \n",
    "\n",
    "# load dataset\n",
    "df_meta <- read_tsv(paste(\"../input/\",file_name,\"_CellMetainfo_table.tsv\", sep=\"\"))\n",
    "h5_data <- hdf5r::H5File$new(paste(\"../input/\",file_name,\"_expression.h5\", sep=\"\"), mode = 'r')\n",
    "\n",
    "feature_matrix <- Matrix::sparseMatrix(\n",
    "  i = h5_data[['matrix/indices']][],\n",
    "  p = h5_data[['matrix/indptr']][],\n",
    "  x = h5_data[['matrix/data']][],\n",
    "  dimnames = list(\n",
    "    h5_data[['matrix/features/name']][],\n",
    "    h5_data[['matrix/barcodes']][]\n",
    "  ),\n",
    "  dims = h5_data[['matrix/shape']][],\n",
    "  index1 = FALSE\n",
    ")\n",
    "rm(h5_data) # for saving memory\n",
    "\n",
    "\n",
    "m_subset <- feature_matrix[rownames(feature_matrix) %in% keep_rows, ] # Extract rows from matrix\n",
    "rm(feature_matrix)\n",
    "\n",
    "df_mat <- as.data.frame(as.matrix(m_subset))\n",
    "df_mat <- as_tibble(t(df_mat), rownames=\"Cell\")\n",
    "\n",
    "df <- df_meta %>% left_join(., df_mat)\n",
    "head(df)\n",
    "\n",
    "# to save files\n",
    "write.csv(df, paste(\"../output/\",file_name,\".csv\", sep=\"\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "1f2a018e-4c45-4c45-94e1-ef056d99a552",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1mRows: \u001b[22m\u001b[34m40218\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m11\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (8): Cell, Celltype (malignancy), Celltype (major-lineage), Celltype (mi...\n",
      "\u001b[32mdbl\u001b[39m (3): UMAP_1, UMAP_2, Cluster\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Cell)`\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 6 × 18</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Cell</th><th scope=col>UMAP_1</th><th scope=col>UMAP_2</th><th scope=col>Celltype (malignancy)</th><th scope=col>Celltype (major-lineage)</th><th scope=col>Celltype (minor-lineage)</th><th scope=col>Celltype (original)</th><th scope=col>Cluster</th><th scope=col>DbCluster</th><th scope=col>Source</th><th scope=col>Stage</th><th scope=col>TIGIT</th><th scope=col>HLA-G</th><th scope=col>CD226</th><th scope=col>PVR</th><th scope=col>LILRB2</th><th scope=col>LILRB1</th><th scope=col>LILRB4</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>AGTTCTACGCATAC_1</td><td> 1.842451</td><td>-16.59187</td><td>Others</td><td>Alveolar</td><td>Alveolar</td><td>Alveolar </td><td>13</td><td>Alveolar </td><td>Tumor</td><td>Primary</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>ATACCGGATCGTGA_1</td><td> 1.753730</td><td>-16.29929</td><td>Others</td><td>Alveolar</td><td>Alveolar</td><td>Alveolar </td><td>13</td><td>Alveolar </td><td>Tumor</td><td>Primary</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>GAACGTTGATGGTC_1</td><td> 1.481534</td><td>-15.80473</td><td>Others</td><td>Alveolar</td><td>Alveolar</td><td>Alveolar </td><td>13</td><td>Alveolar </td><td>Tumor</td><td>Primary</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>GCACGGTGTAACCG_1</td><td> 3.887330</td><td>-14.12924</td><td>Others</td><td>Alveolar</td><td>Alveolar</td><td>Alveolar </td><td>13</td><td>Alveolar </td><td>Tumor</td><td>Primary</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>TCAGACGAGCAAGG_1</td><td> 1.627641</td><td>-16.31171</td><td>Others</td><td>Alveolar</td><td>Alveolar</td><td>Alveolar </td><td>13</td><td>Alveolar </td><td>Tumor</td><td>Primary</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>GCGCACGACAGATC_3</td><td>-7.192630</td><td>-15.41526</td><td>Others</td><td>Alveolar</td><td>Alveolar</td><td>Malignant</td><td>13</td><td>Malignant</td><td>Tumor</td><td>Primary</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 6 × 18\n",
       "\\begin{tabular}{llllllllllllllllll}\n",
       " Cell & UMAP\\_1 & UMAP\\_2 & Celltype (malignancy) & Celltype (major-lineage) & Celltype (minor-lineage) & Celltype (original) & Cluster & DbCluster & Source & Stage & TIGIT & HLA-G & CD226 & PVR & LILRB2 & LILRB1 & LILRB4\\\\\n",
       " <chr> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <chr> & <dbl> & <chr> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t AGTTCTACGCATAC\\_1 &  1.842451 & -16.59187 & Others & Alveolar & Alveolar & Alveolar  & 13 & Alveolar  & Tumor & Primary & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\t ATACCGGATCGTGA\\_1 &  1.753730 & -16.29929 & Others & Alveolar & Alveolar & Alveolar  & 13 & Alveolar  & Tumor & Primary & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\t GAACGTTGATGGTC\\_1 &  1.481534 & -15.80473 & Others & Alveolar & Alveolar & Alveolar  & 13 & Alveolar  & Tumor & Primary & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\t GCACGGTGTAACCG\\_1 &  3.887330 & -14.12924 & Others & Alveolar & Alveolar & Alveolar  & 13 & Alveolar  & Tumor & Primary & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\t TCAGACGAGCAAGG\\_1 &  1.627641 & -16.31171 & Others & Alveolar & Alveolar & Alveolar  & 13 & Alveolar  & Tumor & Primary & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\t GCGCACGACAGATC\\_3 & -7.192630 & -15.41526 & Others & Alveolar & Alveolar & Malignant & 13 & Malignant & Tumor & Primary & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 6 × 18\n",
       "\n",
       "| Cell &lt;chr&gt; | UMAP_1 &lt;dbl&gt; | UMAP_2 &lt;dbl&gt; | Celltype (malignancy) &lt;chr&gt; | Celltype (major-lineage) &lt;chr&gt; | Celltype (minor-lineage) &lt;chr&gt; | Celltype (original) &lt;chr&gt; | Cluster &lt;dbl&gt; | DbCluster &lt;chr&gt; | Source &lt;chr&gt; | Stage &lt;chr&gt; | TIGIT &lt;dbl&gt; | HLA-G &lt;dbl&gt; | CD226 &lt;dbl&gt; | PVR &lt;dbl&gt; | LILRB2 &lt;dbl&gt; | LILRB1 &lt;dbl&gt; | LILRB4 &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| AGTTCTACGCATAC_1 |  1.842451 | -16.59187 | Others | Alveolar | Alveolar | Alveolar  | 13 | Alveolar  | Tumor | Primary | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "| ATACCGGATCGTGA_1 |  1.753730 | -16.29929 | Others | Alveolar | Alveolar | Alveolar  | 13 | Alveolar  | Tumor | Primary | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "| GAACGTTGATGGTC_1 |  1.481534 | -15.80473 | Others | Alveolar | Alveolar | Alveolar  | 13 | Alveolar  | Tumor | Primary | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "| GCACGGTGTAACCG_1 |  3.887330 | -14.12924 | Others | Alveolar | Alveolar | Alveolar  | 13 | Alveolar  | Tumor | Primary | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "| TCAGACGAGCAAGG_1 |  1.627641 | -16.31171 | Others | Alveolar | Alveolar | Alveolar  | 13 | Alveolar  | Tumor | Primary | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "| GCGCACGACAGATC_3 | -7.192630 | -15.41526 | Others | Alveolar | Alveolar | Malignant | 13 | Malignant | Tumor | Primary | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "  Cell             UMAP_1    UMAP_2    Celltype (malignancy)\n",
       "1 AGTTCTACGCATAC_1  1.842451 -16.59187 Others               \n",
       "2 ATACCGGATCGTGA_1  1.753730 -16.29929 Others               \n",
       "3 GAACGTTGATGGTC_1  1.481534 -15.80473 Others               \n",
       "4 GCACGGTGTAACCG_1  3.887330 -14.12924 Others               \n",
       "5 TCAGACGAGCAAGG_1  1.627641 -16.31171 Others               \n",
       "6 GCGCACGACAGATC_3 -7.192630 -15.41526 Others               \n",
       "  Celltype (major-lineage) Celltype (minor-lineage) Celltype (original) Cluster\n",
       "1 Alveolar                 Alveolar                 Alveolar            13     \n",
       "2 Alveolar                 Alveolar                 Alveolar            13     \n",
       "3 Alveolar                 Alveolar                 Alveolar            13     \n",
       "4 Alveolar                 Alveolar                 Alveolar            13     \n",
       "5 Alveolar                 Alveolar                 Alveolar            13     \n",
       "6 Alveolar                 Alveolar                 Malignant           13     \n",
       "  DbCluster Source Stage   TIGIT HLA-G CD226 PVR LILRB2 LILRB1 LILRB4\n",
       "1 Alveolar  Tumor  Primary 0     0     0     0   0      0      0     \n",
       "2 Alveolar  Tumor  Primary 0     0     0     0   0      0      0     \n",
       "3 Alveolar  Tumor  Primary 0     0     0     0   0      0      0     \n",
       "4 Alveolar  Tumor  Primary 0     0     0     0   0      0      0     \n",
       "5 Alveolar  Tumor  Primary 0     0     0     0   0      0      0     \n",
       "6 Malignant Tumor  Primary 0     0     0     0   0      0      0     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name <- \"NSCLC_EMTAB6149\"\n",
    "# Specify rows to keep\n",
    "keep_rows <- c(\"TIGIT\",\"PVR\",\"CD226\",\"PVRL2\",\"LILRB1\", \"LILRB2\", \"LILRB4\", \"HLA-G\")         \n",
    "\n",
    "# load dataset\n",
    "df_meta <- read_tsv(paste(\"../input/\",file_name,\"_CellMetainfo_table.tsv\", sep=\"\"))\n",
    "h5_data <- hdf5r::H5File$new(paste(\"../input/\",file_name,\"_expression.h5\", sep=\"\"), mode = 'r')\n",
    "\n",
    "feature_matrix <- Matrix::sparseMatrix(\n",
    "  i = h5_data[['matrix/indices']][],\n",
    "  p = h5_data[['matrix/indptr']][],\n",
    "  x = h5_data[['matrix/data']][],\n",
    "  dimnames = list(\n",
    "    h5_data[['matrix/features/name']][],\n",
    "    h5_data[['matrix/barcodes']][]\n",
    "  ),\n",
    "  dims = h5_data[['matrix/shape']][],\n",
    "  index1 = FALSE\n",
    ")\n",
    "rm(h5_data) # for saving memory\n",
    "\n",
    "\n",
    "m_subset <- feature_matrix[rownames(feature_matrix) %in% keep_rows, ] # Extract rows from matrix\n",
    "rm(feature_matrix)\n",
    "\n",
    "df_mat <- as.data.frame(as.matrix(m_subset))\n",
    "df_mat <- as_tibble(t(df_mat), rownames=\"Cell\")\n",
    "\n",
    "df <- df_meta %>% left_join(., df_mat)\n",
    "head(df)\n",
    "\n",
    "# to save files\n",
    "write.csv(df, paste(\"../output/\",file_name,\".csv\", sep=\"\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "55b364df-156a-426b-b250-385f063b91b5",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1mRows: \u001b[22m\u001b[34m89471\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m13\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (10): Cell, Celltype (malignancy), Celltype (major-lineage), Celltype (m...\n",
      "\u001b[32mdbl\u001b[39m  (3): UMAP_1, UMAP_2, Cluster\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Cell)`\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 6 × 21</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Cell</th><th scope=col>UMAP_1</th><th scope=col>UMAP_2</th><th scope=col>Celltype (malignancy)</th><th scope=col>Celltype (major-lineage)</th><th scope=col>Celltype (minor-lineage)</th><th scope=col>Celltype (original)</th><th scope=col>Cluster</th><th scope=col>Subtype</th><th scope=col>Celltype_major</th><th scope=col>⋯</th><th scope=col>Celltype_minor</th><th scope=col>Patient</th><th scope=col>TIGIT</th><th scope=col>HLA-G</th><th scope=col>CD226</th><th scope=col>PVR</th><th scope=col>PVRL2</th><th scope=col>LILRB2</th><th scope=col>LILRB1</th><th scope=col>LILRB4</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>⋯</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>CID4495_AAGCCGCAGGCATGGT</td><td>-7.615879</td><td>-6.630866</td><td>Immune cells</td><td>B</td><td>B</td><td>Endothelial</td><td>8</td><td>TNBC</td><td>Endothelial</td><td>⋯</td><td>Endothelial_Endothelial ACKR1+</td><td>CID4495</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>2.761641</td></tr>\n",
       "\t<tr><td>CID4495_GATCGATTCACGAAGG</td><td>-7.629208</td><td>-6.569772</td><td>Immune cells</td><td>B</td><td>B</td><td>Endothelial</td><td>8</td><td>TNBC</td><td>Endothelial</td><td>⋯</td><td>Endothelial_Endothelial ACKR1+</td><td>CID4495</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>2.109138</td></tr>\n",
       "\t<tr><td>CID4495_CATCAAGCAGTGACAG</td><td>-7.570159</td><td>-6.324082</td><td>Immune cells</td><td>B</td><td>B</td><td>CAFs       </td><td>8</td><td>TNBC</td><td>CAFs       </td><td>⋯</td><td>CAFs_Inflammatory-CAFs        </td><td>CID4495</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>3.010868</td></tr>\n",
       "\t<tr><td>CID4495_CACCTTGCACCGAAAG</td><td>-9.622888</td><td>-3.392681</td><td>Immune cells</td><td>B</td><td>B</td><td>CAFs       </td><td>8</td><td>TNBC</td><td>CAFs       </td><td>⋯</td><td>CAFs_Inflammatory-CAFs        </td><td>CID4495</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0.000000</td></tr>\n",
       "\t<tr><td>CID4495_TGCGCAGGTTGGTAAA</td><td>-7.676848</td><td>-6.092321</td><td>Immune cells</td><td>B</td><td>B</td><td>CAFs       </td><td>8</td><td>TNBC</td><td>CAFs       </td><td>⋯</td><td>CAFs_Inflammatory-CAFs        </td><td>CID4495</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>2.286632</td></tr>\n",
       "\t<tr><td>CID4495_TGGGAAGCACTCAGGC</td><td>-7.651750</td><td>-6.488243</td><td>Immune cells</td><td>B</td><td>B</td><td>CAFs       </td><td>8</td><td>TNBC</td><td>CAFs       </td><td>⋯</td><td>CAFs_Inflammatory-CAFs        </td><td>CID4495</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0.000000</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 6 × 21\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " Cell & UMAP\\_1 & UMAP\\_2 & Celltype (malignancy) & Celltype (major-lineage) & Celltype (minor-lineage) & Celltype (original) & Cluster & Subtype & Celltype\\_major & ⋯ & Celltype\\_minor & Patient & TIGIT & HLA-G & CD226 & PVR & PVRL2 & LILRB2 & LILRB1 & LILRB4\\\\\n",
       " <chr> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <chr> & <dbl> & <chr> & <chr> & ⋯ & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t CID4495\\_AAGCCGCAGGCATGGT & -7.615879 & -6.630866 & Immune cells & B & B & Endothelial & 8 & TNBC & Endothelial & ⋯ & Endothelial\\_Endothelial ACKR1+ & CID4495 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 2.761641\\\\\n",
       "\t CID4495\\_GATCGATTCACGAAGG & -7.629208 & -6.569772 & Immune cells & B & B & Endothelial & 8 & TNBC & Endothelial & ⋯ & Endothelial\\_Endothelial ACKR1+ & CID4495 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 2.109138\\\\\n",
       "\t CID4495\\_CATCAAGCAGTGACAG & -7.570159 & -6.324082 & Immune cells & B & B & CAFs        & 8 & TNBC & CAFs        & ⋯ & CAFs\\_Inflammatory-CAFs         & CID4495 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 3.010868\\\\\n",
       "\t CID4495\\_CACCTTGCACCGAAAG & -9.622888 & -3.392681 & Immune cells & B & B & CAFs        & 8 & TNBC & CAFs        & ⋯ & CAFs\\_Inflammatory-CAFs         & CID4495 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0.000000\\\\\n",
       "\t CID4495\\_TGCGCAGGTTGGTAAA & -7.676848 & -6.092321 & Immune cells & B & B & CAFs        & 8 & TNBC & CAFs        & ⋯ & CAFs\\_Inflammatory-CAFs         & CID4495 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 2.286632\\\\\n",
       "\t CID4495\\_TGGGAAGCACTCAGGC & -7.651750 & -6.488243 & Immune cells & B & B & CAFs        & 8 & TNBC & CAFs        & ⋯ & CAFs\\_Inflammatory-CAFs         & CID4495 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0.000000\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 6 × 21\n",
       "\n",
       "| Cell &lt;chr&gt; | UMAP_1 &lt;dbl&gt; | UMAP_2 &lt;dbl&gt; | Celltype (malignancy) &lt;chr&gt; | Celltype (major-lineage) &lt;chr&gt; | Celltype (minor-lineage) &lt;chr&gt; | Celltype (original) &lt;chr&gt; | Cluster &lt;dbl&gt; | Subtype &lt;chr&gt; | Celltype_major &lt;chr&gt; | ⋯ ⋯ | Celltype_minor &lt;chr&gt; | Patient &lt;chr&gt; | TIGIT &lt;dbl&gt; | HLA-G &lt;dbl&gt; | CD226 &lt;dbl&gt; | PVR &lt;dbl&gt; | PVRL2 &lt;dbl&gt; | LILRB2 &lt;dbl&gt; | LILRB1 &lt;dbl&gt; | LILRB4 &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| CID4495_AAGCCGCAGGCATGGT | -7.615879 | -6.630866 | Immune cells | B | B | Endothelial | 8 | TNBC | Endothelial | ⋯ | Endothelial_Endothelial ACKR1+ | CID4495 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 2.761641 |\n",
       "| CID4495_GATCGATTCACGAAGG | -7.629208 | -6.569772 | Immune cells | B | B | Endothelial | 8 | TNBC | Endothelial | ⋯ | Endothelial_Endothelial ACKR1+ | CID4495 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 2.109138 |\n",
       "| CID4495_CATCAAGCAGTGACAG | -7.570159 | -6.324082 | Immune cells | B | B | CAFs        | 8 | TNBC | CAFs        | ⋯ | CAFs_Inflammatory-CAFs         | CID4495 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 3.010868 |\n",
       "| CID4495_CACCTTGCACCGAAAG | -9.622888 | -3.392681 | Immune cells | B | B | CAFs        | 8 | TNBC | CAFs        | ⋯ | CAFs_Inflammatory-CAFs         | CID4495 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0.000000 |\n",
       "| CID4495_TGCGCAGGTTGGTAAA | -7.676848 | -6.092321 | Immune cells | B | B | CAFs        | 8 | TNBC | CAFs        | ⋯ | CAFs_Inflammatory-CAFs         | CID4495 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 2.286632 |\n",
       "| CID4495_TGGGAAGCACTCAGGC | -7.651750 | -6.488243 | Immune cells | B | B | CAFs        | 8 | TNBC | CAFs        | ⋯ | CAFs_Inflammatory-CAFs         | CID4495 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0.000000 |\n",
       "\n"
      ],
      "text/plain": [
       "  Cell                     UMAP_1    UMAP_2    Celltype (malignancy)\n",
       "1 CID4495_AAGCCGCAGGCATGGT -7.615879 -6.630866 Immune cells         \n",
       "2 CID4495_GATCGATTCACGAAGG -7.629208 -6.569772 Immune cells         \n",
       "3 CID4495_CATCAAGCAGTGACAG -7.570159 -6.324082 Immune cells         \n",
       "4 CID4495_CACCTTGCACCGAAAG -9.622888 -3.392681 Immune cells         \n",
       "5 CID4495_TGCGCAGGTTGGTAAA -7.676848 -6.092321 Immune cells         \n",
       "6 CID4495_TGGGAAGCACTCAGGC -7.651750 -6.488243 Immune cells         \n",
       "  Celltype (major-lineage) Celltype (minor-lineage) Celltype (original) Cluster\n",
       "1 B                        B                        Endothelial         8      \n",
       "2 B                        B                        Endothelial         8      \n",
       "3 B                        B                        CAFs                8      \n",
       "4 B                        B                        CAFs                8      \n",
       "5 B                        B                        CAFs                8      \n",
       "6 B                        B                        CAFs                8      \n",
       "  Subtype Celltype_major ⋯ Celltype_minor                 Patient TIGIT HLA-G\n",
       "1 TNBC    Endothelial    ⋯ Endothelial_Endothelial ACKR1+ CID4495 0     0    \n",
       "2 TNBC    Endothelial    ⋯ Endothelial_Endothelial ACKR1+ CID4495 0     0    \n",
       "3 TNBC    CAFs           ⋯ CAFs_Inflammatory-CAFs         CID4495 0     0    \n",
       "4 TNBC    CAFs           ⋯ CAFs_Inflammatory-CAFs         CID4495 0     0    \n",
       "5 TNBC    CAFs           ⋯ CAFs_Inflammatory-CAFs         CID4495 0     0    \n",
       "6 TNBC    CAFs           ⋯ CAFs_Inflammatory-CAFs         CID4495 0     0    \n",
       "  CD226 PVR PVRL2 LILRB2 LILRB1 LILRB4  \n",
       "1 0     0   0     0      0      2.761641\n",
       "2 0     0   0     0      0      2.109138\n",
       "3 0     0   0     0      0      3.010868\n",
       "4 0     0   0     0      0      0.000000\n",
       "5 0     0   0     0      0      2.286632\n",
       "6 0     0   0     0      0      0.000000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name <- \"BRCA_GSE176078\"\n",
    "# Specify rows to keep\n",
    "keep_rows <- c(\"TIGIT\",\"PVR\",\"CD226\",\"PVRL2\",\"LILRB1\", \"LILRB2\", \"LILRB4\", \"HLA-G\")         \n",
    "\n",
    "# load dataset\n",
    "df_meta <- read_tsv(paste(\"../input/\",file_name,\"_CellMetainfo_table.tsv\", sep=\"\"))\n",
    "h5_data <- hdf5r::H5File$new(paste(\"../input/\",file_name,\"_expression.h5\", sep=\"\"), mode = 'r')\n",
    "\n",
    "feature_matrix <- Matrix::sparseMatrix(\n",
    "  i = h5_data[['matrix/indices']][],\n",
    "  p = h5_data[['matrix/indptr']][],\n",
    "  x = h5_data[['matrix/data']][],\n",
    "  dimnames = list(\n",
    "    h5_data[['matrix/features/name']][],\n",
    "    h5_data[['matrix/barcodes']][]\n",
    "  ),\n",
    "  dims = h5_data[['matrix/shape']][],\n",
    "  index1 = FALSE\n",
    ")\n",
    "rm(h5_data) # for saving memory\n",
    "\n",
    "\n",
    "m_subset <- feature_matrix[rownames(feature_matrix) %in% keep_rows, ] # Extract rows from matrix\n",
    "rm(feature_matrix)\n",
    "\n",
    "df_mat <- as.data.frame(as.matrix(m_subset))\n",
    "df_mat <- as_tibble(t(df_mat), rownames=\"Cell\")\n",
    "\n",
    "df <- df_meta %>% left_join(., df_mat)\n",
    "head(df)\n",
    "\n",
    "# to save files\n",
    "write.csv(df, paste(\"../output/\",file_name,\".csv\", sep=\"\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f7f6de21-a217-418f-81dc-2b5a0c424d2d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1mRows: \u001b[22m\u001b[34m38348\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m15\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (12): Cell, Celltype (malignancy), Celltype (major-lineage), Celltype (m...\n",
      "\u001b[32mdbl\u001b[39m  (3): UMAP_1, UMAP_2, Cluster\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Cell)`\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 6 × 22</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Cell</th><th scope=col>UMAP_1</th><th scope=col>UMAP_2</th><th scope=col>Celltype (malignancy)</th><th scope=col>Celltype (major-lineage)</th><th scope=col>Celltype (minor-lineage)</th><th scope=col>Celltype (original)</th><th scope=col>Cluster</th><th scope=col>PredictionRefined</th><th scope=col>Sample</th><th scope=col>⋯</th><th scope=col>Patient</th><th scope=col>Gender</th><th scope=col>Stage</th><th scope=col>CD226</th><th scope=col>HLA-G</th><th scope=col>LILRB1</th><th scope=col>LILRB2</th><th scope=col>LILRB4</th><th scope=col>PVR</th><th scope=col>TIGIT</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>⋯</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>AML328-D113_ATTAAGCATTTC</td><td>-3.304277</td><td>-4.353379</td><td>Immune cells</td><td>B</td><td>B</td><td>B</td><td>18</td><td>normal</td><td>AML328-D113</td><td>⋯</td><td>AML328</td><td>Female</td><td>Primary</td><td>0</td><td>0</td><td>1.924949</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>AML328-D113_GAAATGCCCCAT</td><td>-3.453818</td><td>-4.270767</td><td>Immune cells</td><td>B</td><td>B</td><td>B</td><td>18</td><td>normal</td><td>AML328-D113</td><td>⋯</td><td>AML328</td><td>Female</td><td>Primary</td><td>0</td><td>0</td><td>0.000000</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>AML328-D113_GGGTGGCGAAAC</td><td>-3.338420</td><td>-4.399727</td><td>Immune cells</td><td>B</td><td>B</td><td>B</td><td>18</td><td>normal</td><td>AML328-D113</td><td>⋯</td><td>AML328</td><td>Female</td><td>Primary</td><td>0</td><td>0</td><td>0.000000</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>AML328-D113_TAACTAAGCTAC</td><td>-3.465750</td><td>-4.229091</td><td>Immune cells</td><td>B</td><td>B</td><td>B</td><td>18</td><td>normal</td><td>AML328-D113</td><td>⋯</td><td>AML328</td><td>Female</td><td>Primary</td><td>0</td><td>0</td><td>0.000000</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>AML328-D113_TCGTTAGTTATT</td><td>-3.474983</td><td>-4.395324</td><td>Immune cells</td><td>B</td><td>B</td><td>B</td><td>18</td><td>normal</td><td>AML328-D113</td><td>⋯</td><td>AML328</td><td>Female</td><td>Primary</td><td>0</td><td>0</td><td>1.305719</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>AML328-D113_TCTAGATTCCGA</td><td>-3.432360</td><td>-4.451101</td><td>Immune cells</td><td>B</td><td>B</td><td>B</td><td>18</td><td>normal</td><td>AML328-D113</td><td>⋯</td><td>AML328</td><td>Female</td><td>Primary</td><td>0</td><td>0</td><td>0.000000</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 6 × 22\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " Cell & UMAP\\_1 & UMAP\\_2 & Celltype (malignancy) & Celltype (major-lineage) & Celltype (minor-lineage) & Celltype (original) & Cluster & PredictionRefined & Sample & ⋯ & Patient & Gender & Stage & CD226 & HLA-G & LILRB1 & LILRB2 & LILRB4 & PVR & TIGIT\\\\\n",
       " <chr> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <chr> & <dbl> & <chr> & <chr> & ⋯ & <chr> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t AML328-D113\\_ATTAAGCATTTC & -3.304277 & -4.353379 & Immune cells & B & B & B & 18 & normal & AML328-D113 & ⋯ & AML328 & Female & Primary & 0 & 0 & 1.924949 & 0 & 0 & 0 & 0\\\\\n",
       "\t AML328-D113\\_GAAATGCCCCAT & -3.453818 & -4.270767 & Immune cells & B & B & B & 18 & normal & AML328-D113 & ⋯ & AML328 & Female & Primary & 0 & 0 & 0.000000 & 0 & 0 & 0 & 0\\\\\n",
       "\t AML328-D113\\_GGGTGGCGAAAC & -3.338420 & -4.399727 & Immune cells & B & B & B & 18 & normal & AML328-D113 & ⋯ & AML328 & Female & Primary & 0 & 0 & 0.000000 & 0 & 0 & 0 & 0\\\\\n",
       "\t AML328-D113\\_TAACTAAGCTAC & -3.465750 & -4.229091 & Immune cells & B & B & B & 18 & normal & AML328-D113 & ⋯ & AML328 & Female & Primary & 0 & 0 & 0.000000 & 0 & 0 & 0 & 0\\\\\n",
       "\t AML328-D113\\_TCGTTAGTTATT & -3.474983 & -4.395324 & Immune cells & B & B & B & 18 & normal & AML328-D113 & ⋯ & AML328 & Female & Primary & 0 & 0 & 1.305719 & 0 & 0 & 0 & 0\\\\\n",
       "\t AML328-D113\\_TCTAGATTCCGA & -3.432360 & -4.451101 & Immune cells & B & B & B & 18 & normal & AML328-D113 & ⋯ & AML328 & Female & Primary & 0 & 0 & 0.000000 & 0 & 0 & 0 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 6 × 22\n",
       "\n",
       "| Cell &lt;chr&gt; | UMAP_1 &lt;dbl&gt; | UMAP_2 &lt;dbl&gt; | Celltype (malignancy) &lt;chr&gt; | Celltype (major-lineage) &lt;chr&gt; | Celltype (minor-lineage) &lt;chr&gt; | Celltype (original) &lt;chr&gt; | Cluster &lt;dbl&gt; | PredictionRefined &lt;chr&gt; | Sample &lt;chr&gt; | ⋯ ⋯ | Patient &lt;chr&gt; | Gender &lt;chr&gt; | Stage &lt;chr&gt; | CD226 &lt;dbl&gt; | HLA-G &lt;dbl&gt; | LILRB1 &lt;dbl&gt; | LILRB2 &lt;dbl&gt; | LILRB4 &lt;dbl&gt; | PVR &lt;dbl&gt; | TIGIT &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| AML328-D113_ATTAAGCATTTC | -3.304277 | -4.353379 | Immune cells | B | B | B | 18 | normal | AML328-D113 | ⋯ | AML328 | Female | Primary | 0 | 0 | 1.924949 | 0 | 0 | 0 | 0 |\n",
       "| AML328-D113_GAAATGCCCCAT | -3.453818 | -4.270767 | Immune cells | B | B | B | 18 | normal | AML328-D113 | ⋯ | AML328 | Female | Primary | 0 | 0 | 0.000000 | 0 | 0 | 0 | 0 |\n",
       "| AML328-D113_GGGTGGCGAAAC | -3.338420 | -4.399727 | Immune cells | B | B | B | 18 | normal | AML328-D113 | ⋯ | AML328 | Female | Primary | 0 | 0 | 0.000000 | 0 | 0 | 0 | 0 |\n",
       "| AML328-D113_TAACTAAGCTAC | -3.465750 | -4.229091 | Immune cells | B | B | B | 18 | normal | AML328-D113 | ⋯ | AML328 | Female | Primary | 0 | 0 | 0.000000 | 0 | 0 | 0 | 0 |\n",
       "| AML328-D113_TCGTTAGTTATT | -3.474983 | -4.395324 | Immune cells | B | B | B | 18 | normal | AML328-D113 | ⋯ | AML328 | Female | Primary | 0 | 0 | 1.305719 | 0 | 0 | 0 | 0 |\n",
       "| AML328-D113_TCTAGATTCCGA | -3.432360 | -4.451101 | Immune cells | B | B | B | 18 | normal | AML328-D113 | ⋯ | AML328 | Female | Primary | 0 | 0 | 0.000000 | 0 | 0 | 0 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "  Cell                     UMAP_1    UMAP_2    Celltype (malignancy)\n",
       "1 AML328-D113_ATTAAGCATTTC -3.304277 -4.353379 Immune cells         \n",
       "2 AML328-D113_GAAATGCCCCAT -3.453818 -4.270767 Immune cells         \n",
       "3 AML328-D113_GGGTGGCGAAAC -3.338420 -4.399727 Immune cells         \n",
       "4 AML328-D113_TAACTAAGCTAC -3.465750 -4.229091 Immune cells         \n",
       "5 AML328-D113_TCGTTAGTTATT -3.474983 -4.395324 Immune cells         \n",
       "6 AML328-D113_TCTAGATTCCGA -3.432360 -4.451101 Immune cells         \n",
       "  Celltype (major-lineage) Celltype (minor-lineage) Celltype (original) Cluster\n",
       "1 B                        B                        B                   18     \n",
       "2 B                        B                        B                   18     \n",
       "3 B                        B                        B                   18     \n",
       "4 B                        B                        B                   18     \n",
       "5 B                        B                        B                   18     \n",
       "6 B                        B                        B                   18     \n",
       "  PredictionRefined Sample      ⋯ Patient Gender Stage   CD226 HLA-G LILRB1  \n",
       "1 normal            AML328-D113 ⋯ AML328  Female Primary 0     0     1.924949\n",
       "2 normal            AML328-D113 ⋯ AML328  Female Primary 0     0     0.000000\n",
       "3 normal            AML328-D113 ⋯ AML328  Female Primary 0     0     0.000000\n",
       "4 normal            AML328-D113 ⋯ AML328  Female Primary 0     0     0.000000\n",
       "5 normal            AML328-D113 ⋯ AML328  Female Primary 0     0     1.305719\n",
       "6 normal            AML328-D113 ⋯ AML328  Female Primary 0     0     0.000000\n",
       "  LILRB2 LILRB4 PVR TIGIT\n",
       "1 0      0      0   0    \n",
       "2 0      0      0   0    \n",
       "3 0      0      0   0    \n",
       "4 0      0      0   0    \n",
       "5 0      0      0   0    \n",
       "6 0      0      0   0    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name <- \"AML_GSE116256\"\n",
    "# Specify rows to keep\n",
    "keep_rows <- c(\"TIGIT\",\"PVR\",\"CD226\",\"PVRL2\",\"LILRB1\", \"LILRB2\", \"LILRB4\", \"HLA-G\")         \n",
    "\n",
    "# load dataset\n",
    "df_meta <- read_tsv(paste(\"../input/\",file_name,\"_CellMetainfo_table.tsv\", sep=\"\"))\n",
    "h5_data <- hdf5r::H5File$new(paste(\"../input/\",file_name,\"_expression.h5\", sep=\"\"), mode = 'r')\n",
    "\n",
    "feature_matrix <- Matrix::sparseMatrix(\n",
    "  i = h5_data[['matrix/indices']][],\n",
    "  p = h5_data[['matrix/indptr']][],\n",
    "  x = h5_data[['matrix/data']][],\n",
    "  dimnames = list(\n",
    "    h5_data[['matrix/features/name']][],\n",
    "    h5_data[['matrix/barcodes']][]\n",
    "  ),\n",
    "  dims = h5_data[['matrix/shape']][],\n",
    "  index1 = FALSE\n",
    ")\n",
    "rm(h5_data) # for saving memory\n",
    "\n",
    "\n",
    "m_subset <- feature_matrix[rownames(feature_matrix) %in% keep_rows, ] # Extract rows from matrix\n",
    "rm(feature_matrix)\n",
    "\n",
    "df_mat <- as.data.frame(as.matrix(m_subset))\n",
    "df_mat <- as_tibble(t(df_mat), rownames=\"Cell\")\n",
    "\n",
    "df <- df_meta %>% left_join(., df_mat)\n",
    "head(df)\n",
    "\n",
    "# to save files\n",
    "write.csv(df, paste(\"../output/\",file_name,\".csv\", sep=\"\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d083c6e-0f15-4be4-88a2-81c3fe5e86f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_name <- \"AML_GSE116256\"\n",
    "# Specify rows to keep\n",
    "keep_rows <- c(\"TIGIT\",\"PVR\",\"CD226\",\"PVRL2\",\"LILRB1\", \"LILRB2\", \"LILRB4\", \"HLA-G\")         \n",
    "\n",
    "# load dataset\n",
    "df_meta <- read_tsv(paste(\"../input/\",file_name,\"_CellMetainfo_table.tsv\", sep=\"\"))\n",
    "h5_data <- hdf5r::H5File$new(paste(\"../input/\",file_name,\"_expression.h5\", sep=\"\"), mode = 'r')\n",
    "\n",
    "feature_matrix <- Matrix::sparseMatrix(\n",
    "  i = h5_data[['matrix/indices']][],\n",
    "  p = h5_data[['matrix/indptr']][],\n",
    "  x = h5_data[['matrix/data']][],\n",
    "  dimnames = list(\n",
    "    h5_data[['matrix/features/name']][],\n",
    "    h5_data[['matrix/barcodes']][]\n",
    "  ),\n",
    "  dims = h5_data[['matrix/shape']][],\n",
    "  index1 = FALSE\n",
    ")\n",
    "rm(h5_data) # for saving memory\n",
    "\n",
    "\n",
    "m_subset <- feature_matrix[rownames(feature_matrix) %in% keep_rows, ] # Extract rows from matrix\n",
    "rm(feature_matrix)\n",
    "\n",
    "df_mat <- as.data.frame(as.matrix(m_subset))\n",
    "df_mat <- as_tibble(t(df_mat), rownames=\"Cell\")\n",
    "\n",
    "df <- df_meta %>% left_join(., df_mat)\n",
    "head(df)\n",
    "\n",
    "# to save files\n",
    "write.csv(df, paste(\"../output/\",file_name,\".csv\", sep=\"\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
